// @circles-sdk/pathfinder/dist/index.js
import { ethers } from 'ethers';
import { CirclesData, CirclesRpc } from '@circles-sdk/data';
import { CirclesConverter } from '@circles-sdk/utils';

/**
 * Pack a uint16 array into a hex string (big‑endian, no padding).
 */
function packCoordinates(coords) {
    const bytes = new Uint8Array(coords.length * 2);
    coords.forEach((c, i) => {
        const hi = c >> 8;
        const lo = c & 0xff;
        const offset = 2 * i;
        bytes[offset] = hi;
        bytes[offset + 1] = lo;
    });
    return ethers.hexlify(bytes);
}
/**
 * Build a sorted vertex list plus index lookup for quick coordinate mapping.
 */
function transformToFlowVertices(transfers, from, to) {
    const set = new Set([from.toLowerCase(), to.toLowerCase()]);
    transfers.forEach((t) => {
        set.add(t.from.toLowerCase());
        set.add(t.to.toLowerCase());
        set.add(t.tokenOwner.toLowerCase());
    });
    const sorted = [...set].sort((a, b) => {
        const lhs = BigInt(a);
        const rhs = BigInt(b);
        const isLess = lhs < rhs;
        const isGreater = lhs > rhs;
        return isLess ? -1 : isGreater ? 1 : 0;
    });
    const idx = {};
    sorted.forEach((addr, i) => {
        idx[addr] = i;
    });
    return { sorted, idx };
}

/**
 * Create an ABI‑ready FlowMatrix object from a list of TransferSteps.
 */
function createFlowMatrix(from, to, value, transfers) {
    const sender = from.toLowerCase();
    const receiver = to.toLowerCase();
    const { sorted: flowVertices, idx } = transformToFlowVertices(transfers, sender, receiver);
    const flowEdges = transfers.map((t) => {
        const isTerminal = t.to.toLowerCase() === receiver;
        return {
            streamSinkId: isTerminal ? 1 : 0,
            amount: t.value // keep as string – ethers will convert
        };
    });
    // Ensure at least one terminal edge
    const hasTerminalEdge = flowEdges.some((e) => e.streamSinkId === 1);
    if (!hasTerminalEdge) {
        const lastEdgeIndex = transfers
            .map((t) => t.to.toLowerCase())
            .lastIndexOf(receiver);
        const fallbackIndex = lastEdgeIndex === -1 ? flowEdges.length - 1 : lastEdgeIndex;
        flowEdges[fallbackIndex].streamSinkId = 1;
    }
    const termEdgeIds = flowEdges
        .map((e, i) => (e.streamSinkId === 1 ? i : -1))
        .filter((i) => i !== -1);
    const streams = [
        {
            sourceCoordinate: idx[sender],
            flowEdgeIds: termEdgeIds,
            data: new Uint8Array(0)
        }
    ];
    const coords = [];
    transfers.forEach((t) => {
        coords.push(idx[t.tokenOwner.toLowerCase()]);
        coords.push(idx[t.from.toLowerCase()]);
        coords.push(idx[t.to.toLowerCase()]);
    });
    const packedCoordinates = packCoordinates(coords);
    const expected = BigInt(value);
    const terminalSum = flowEdges
        .filter((e) => e.streamSinkId === 1)
        .reduce((sum, e) => sum + BigInt(e.amount.toString()), BigInt(0));
    const isBalanced = terminalSum === expected;
    if (!isBalanced) {
        throw new Error(`Terminal sum ${terminalSum} does not equal expected ${expected}`);
    }
    return {
        flowVertices,
        flowEdges,
        streams,
        packedCoordinates,
        sourceCoordinate: idx[sender]
    };
}

async function getTokenInfoMapFromPath(rpcUrl, transferPath) {
    const tokenInfoMap = new Map();
    const uniqueAddresses = new Set();
    transferPath.transfers.forEach((t) => {
        uniqueAddresses.add(t.tokenOwner.toLowerCase());
    });
    const circlesData = new CirclesData(rpcUrl);
    const batch = await circlesData.getTokenInfoBatch(Array.from(uniqueAddresses));
    batch.forEach((info) => {
        tokenInfoMap.set(info.token.toLowerCase(), info);
    });
    return tokenInfoMap;
}
function getWrappedTokenTotalsFromPath(transferPath, tokenInfoMap) {
    const wrappedEdgeTotals = {};
    transferPath.transfers.forEach((t) => {
        const info = tokenInfoMap.get(t.tokenOwner.toLowerCase());
        const isWrapper = info && info.type.startsWith('CrcV2_ERC20WrapperDeployed');
        if (isWrapper) {
            if (!wrappedEdgeTotals[t.tokenOwner]) {
                wrappedEdgeTotals[t.tokenOwner] = [BigInt(0), info.type];
            }
            wrappedEdgeTotals[t.tokenOwner][0] += BigInt(t.value);
        }
    });
    return wrappedEdgeTotals;
}
function getExpectedUnwrappedTokenTotals(wrappedTotals, tokenInfoMap) {
    const unwrapped = {};
    Object.entries(wrappedTotals).forEach(([wrapperAddr, [total, type]]) => {
        const info = tokenInfoMap.get(wrapperAddr.toLowerCase());
        if (!info)
            return;
        if (type === 'CrcV2_ERC20WrapperDeployed_Demurraged') {
            unwrapped[wrapperAddr] = [total, info.tokenOwner];
        }
        if (type === 'CrcV2_ERC20WrapperDeployed_Inflationary') {
            unwrapped[wrapperAddr] = [CirclesConverter.attoStaticCirclesToAttoCircles(total), info.tokenOwner];
        }
    });
    return unwrapped;
}
function replaceWrappedTokens(path, unwrapped) {
    const rewritten = path.transfers.map((edge) => {
        const unwrap = unwrapped[edge.tokenOwner.toLowerCase()];
        const hasUnwrap = Boolean(unwrap);
        const tokenOwner = hasUnwrap ? unwrap[1] : edge.tokenOwner;
        return { ...edge, tokenOwner };
    });
    return { ...path, transfers: rewritten };
}
function shrinkPathValues(path, sink, retainBps = BigInt(999_999_999_999)) {
    const incomingToSink = new Map();
    const scaled = [];
    const DENOM = BigInt(1_000_000_000_000);
    path.transfers.forEach((edge) => {
        const scaledValue = (BigInt(edge.value) * retainBps) / DENOM;
        const isZero = scaledValue === BigInt(0);
        if (isZero) {
            return; // drop sub‑unit flows
        }
        scaled.push({ ...edge, value: scaledValue.toString() });
        incomingToSink.set(edge.to, (incomingToSink.get(edge.to) ?? BigInt(0)) + scaledValue);
    });
    const maxFlow = sink ? incomingToSink.get(sink.toLowerCase()) ?? BigInt(0) : BigInt(0);
    return {
        maxFlow: maxFlow.toString(),
        transfers: scaled
    };
}
function assertNoNettedFlowMismatch(path, overrideSource, overrideSink) {
    const net = computeNettedFlow(path);
    const { source, sink } = getSourceAndSink(path, overrideSource, overrideSink);
    const endpointsCoincide = source === sink;
    net.forEach((balance, addr) => {
        /* ----------------------------------------------------------------
         * Closed-loop case → every vertex must net to zero
         * -------------------------------------------------------------- */
        if (endpointsCoincide) {
            if (balance !== BigInt(0)) {
                throw new Error(`Vertex ${addr} is unbalanced: ${balance}`);
            }
            return; // done – nothing else to check for this addr
        }
        /* ----------------------------------------------------------------
         * Ordinary DAG case → classic source / sink / intermediate rules
         * -------------------------------------------------------------- */
        const isSource = addr === source;
        const isSink = addr === sink;
        if (isSource && balance >= BigInt(0)) {
            throw new Error(`Source ${addr} should be net negative, got ${balance}`);
        }
        if (isSink && balance <= BigInt(0)) {
            throw new Error(`Sink ${addr} should be net positive, got ${balance}`);
        }
        const isIntermediate = !isSource && !isSink;
        if (isIntermediate && balance !== BigInt(0)) {
            throw new Error(`Vertex ${addr} is unbalanced: ${balance}`);
        }
    });
}
function getSourceAndSink(path, overrideSource, overrideSink) {
    const senders = new Set(path.transfers.map((t) => t.from.toLowerCase()));
    const receivers = new Set(path.transfers.map((t) => t.to.toLowerCase()));
    const source = [...senders].find((a) => !receivers.has(a));
    const sink = [...receivers].find((a) => !senders.has(a));
    if (!(source ?? overrideSource) || !(sink ?? overrideSink)) {
        throw new Error('Could not determine unique source / sink');
    }
    return { source: (source ?? overrideSource), sink: (sink ?? overrideSink) };
}
function computeNettedFlow(path) {
    const net = new Map();
    path.transfers.forEach(({ from, to, value }) => {
        const amount = BigInt(value);
        net.set(from.toLowerCase(), (net.get(from.toLowerCase()) ?? BigInt(0)) - amount);
        net.set(to.toLowerCase(), (net.get(to.toLowerCase()) ?? BigInt(0)) + amount);
    });
    return net;
}

/**
 * Call Circles RPC `circlesV2_findPath` and return the raw path result.
 */
async function findPath(rpcUrl, { from, to, targetFlow, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers = 300 }) {
    const res = await new CirclesRpc(rpcUrl).call('circlesV2_findPath', [{
            Source: from,
            Sink: to,
            TargetFlow: targetFlow,
            WithWrap: useWrappedBalances,
            MaxTransfers: maxTransfers,
            FromTokens: fromTokens,
            ToTokens: toTokens,
            ExcludedFromTokens: excludeFromTokens,
            ExcludedToTokens: excludeToTokens
        }]);
    return res.result;
}
async function findMaxFlow(rpcUrl, { from, to, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers }) {
    const targetFlow = '9999999999999999999999999999999999999';
    const path = await findPath(rpcUrl, {
        from,
        to,
        targetFlow,
        useWrappedBalances,
        fromTokens,
        toTokens,
        excludeFromTokens,
        excludeToTokens,
        maxTransfers
    });
    return BigInt(path.maxFlow);
}

export { assertNoNettedFlowMismatch, computeNettedFlow, createFlowMatrix, findMaxFlow, findPath, getExpectedUnwrappedTokenTotals, getTokenInfoMapFromPath, getWrappedTokenTotalsFromPath, packCoordinates, replaceWrappedTokens, shrinkPathValues, transformToFlowVertices };
