// @circles-sdk/adapter-ethers/dist/index.js
import { Wallet, BrowserProvider } from 'ethers';
import { handleTransactionError } from '@circles-sdk/utils';

class EthersContractRunner {
    sendBatchTransaction;
    address;
}
class PrivateKeyContractRunner {
    provider;
    privateKey;
    constructor(provider, privateKey) {
        this.provider = provider;
        this.privateKey = privateKey;
    }
    _wallet;
    async init() {
        this._wallet = new Wallet(this.privateKey, this.provider);
        this.address = await this._wallet.getAddress();
    }
    ensureWallet() {
        if (!this._wallet) {
            throw new Error('Not initialized');
        }
        return this._wallet;
    }
    address;
    estimateGas = async (tx) => {
        return this.ensureWallet().estimateGas(tx);
    };
    call = async (tx) => {
        return this.ensureWallet().call(tx);
    };
    resolveName = async (name) => {
        return this.ensureWallet().resolveName(name);
    };
    sendTransaction = async (tx) => {
        try {
            return { ...await this.ensureWallet().sendTransaction({ ...tx }) };
        }
        catch (e) {
            handleTransactionError(e);
        }
    };
    sendBatchTransaction = () => {
        if (!this._wallet) {
            throw new Error('Not initialized');
        }
        return new PrivateKeyBatchRun(this._wallet);
    };
}
class PrivateKeyBatchRun {
    provider;
    transactions = [];
    constructor(provider) {
        this.provider = provider;
        // No asynchronous operations in the constructor
    }
    addTransaction(tx) {
        this.transactions.push(tx);
    }
    async run() {
        let lastReceipt = null;
        for (const tx of this.transactions) {
            try {
                const txResponse = await this.provider.sendTransaction(tx);
                lastReceipt = await txResponse.wait();
            }
            catch (error) {
                handleTransactionError(error);
            }
        }
        return lastReceipt;
    }
}
class BrowserProviderContractRunner {
    constructor() {
        const w = window;
        const ethereumObject = w.ethereum;
        if (!ethereumObject) {
            throw new Error('No provider found on window.ethereum');
        }
        this.provider = new BrowserProvider(ethereumObject);
    }
    async init() {
        this.address = await this.provider.getSigner().then(signer => signer.getAddress());
    }
    address;
    provider;
    estimateGas = async (tx) => this.provider.estimateGas(tx);
    call = async (tx) => this.provider.call(tx);
    resolveName = async (name) => this.provider.resolveName(name);
    sendTransaction = async (tx) => {
        try {
            const signer = await this.provider.getSigner();
            return { ...await signer.sendTransaction(tx) };
        }
        catch (e) {
            handleTransactionError(e);
        }
    };
    sendBatchTransaction = () => new BrowserProviderBatchRun(this.provider);
}
class BrowserProviderBatchRun {
    provider;
    transactions = [];
    constructor(provider) {
        this.provider = provider;
        // No asynchronous operations in the constructor
    }
    addTransaction(tx) {
        this.transactions.push(tx);
    }
    async run() {
        const signer = await this.provider.getSigner();
        let lastReceipt = null;
        for (const tx of this.transactions) {
            try {
                const txResponse = await signer.sendTransaction(tx);
                lastReceipt = await txResponse.wait();
            }
            catch (error) {
                handleTransactionError(error);
            }
        }
        return lastReceipt;
    }
}
/**
 * Takes an ethers6 provider and a sdk contract runner and creates a contract runner that can be used with ethers6
 * and thus with the typechahain generated ethers6 contract wrappers.
 */
class SdkContractRunnerWrapper {
    provider;
    address;
    sdkContractRunner;
    /**
     * Creates a new EthersContractRunner. The provider is used to fetch the transaction details and the sdkContractRunner is used to execute the transactions.
     * @param provider The ethers6 provider
     * @param address The address of the account that signs transactions
     * @param sdkContractRunner The sdk contract runner
     */
    constructor(provider, address, sdkContractRunner) {
        this.provider = provider;
        this.address = address;
        this.sdkContractRunner = sdkContractRunner;
    }
    async init() {
    }
    async addressLikeToString(addressLike) {
        if (typeof addressLike === 'string') {
            return addressLike;
        }
        if (addressLike instanceof Promise) {
            return addressLike;
        }
        return addressLike.getAddress();
    }
    bignumberishToBigInt(value) {
        if (!value) {
            return BigInt(0);
        }
        if (typeof value === 'bigint') {
            return value;
        }
        return BigInt(value);
    }
    estimateGas = async (tx) => {
        if (!this.sdkContractRunner.estimateGas) {
            throw new Error('estimateGas not supported');
        }
        if (!tx.to) {
            throw new Error('to is required');
        }
        if (!tx.data && !tx.value) {
            throw new Error('data or value is required');
        }
        return this.sdkContractRunner.estimateGas({
            to: await this.addressLikeToString(tx.to),
            data: tx.data ?? '0x',
            value: this.bignumberishToBigInt(tx.value)
        });
    };
    call = async (tx) => {
        if (!this.sdkContractRunner.call) {
            throw new Error('call not supported');
        }
        if (!tx.to) {
            throw new Error('to is required');
        }
        if (!tx.data && !tx.value) {
            throw new Error('data or value is required');
        }
        return this.sdkContractRunner.call({
            to: await this.addressLikeToString(tx.to),
            data: tx.data ?? '0x',
            value: this.bignumberishToBigInt(tx.value)
        });
    };
    resolveName;
    sendTransaction = async (tx) => {
        if (!this.sdkContractRunner.sendTransaction) {
            throw new Error('sendTransaction not supported');
        }
        if (!tx.to) {
            throw new Error('to is required');
        }
        if (!tx.data && !tx.value) {
            throw new Error('data or value is required');
        }
        try {
            const response = await this.sdkContractRunner.sendTransaction({
                to: await this.addressLikeToString(tx.to),
                data: tx.data ?? '0x',
                value: this.bignumberishToBigInt(tx.value)
            });
            const transactionResponse = await this.provider.getTransaction(response.hash);
            if (!transactionResponse) {
                throw new Error('Transaction not found');
            }
            return { ...transactionResponse };
        }
        catch (e) {
            handleTransactionError(e);
        }
    };
    sendBatchTransaction() {
        if (!this.sdkContractRunner.sendBatchTransaction) {
            throw new Error('sendBatchTransaction not supported');
        }
        return this.sdkContractRunner.sendBatchTransaction();
    }
}

export { BrowserProviderBatchRun, BrowserProviderContractRunner, EthersContractRunner, PrivateKeyBatchRun, PrivateKeyContractRunner, SdkContractRunnerWrapper };
