// @circles-sdk/utils/dist/index.js
import { ethers } from 'ethers';
import multihash from 'multihashes';

// Token type enum
var BaseGroupMintType;
(function (BaseGroupMintType) {
    BaseGroupMintType["ERC1155"] = "ERC1155";
    BaseGroupMintType["INFLATIONARY"] = "TYPE_INFLATIONARY";
    BaseGroupMintType["DEMURRAGE"] = "TYPE_DEMURRAGE";
})(BaseGroupMintType || (BaseGroupMintType = {}));

/**
 * Conversions between demurrage and inflationary units in V2 – bit‑identical with the
 * Solidity reference implementation (ABDK Math 64.64).
 *
 * All fixed‑point math uses ABDK’s 64.64 format encoded in `bigint`.
 */
class CirclesConverter {
    // ───────────────────────────────── constants ─────────────────────────────────
    /** 1.0 in 64.64 representation */
    static ONE_64 = 1n << 64n;
    /** GAMMA = (0.93)^(1/365.25) * 2^64 */
    static GAMMA_64 = 18443079296116538654n;
    /** BETA = 1 / GAMMA * 2^64 */
    static BETA_64 = 18450409579521241655n;
    static SECONDS_PER_DAY = 86400n;
    static INFLATION_DAY_ZERO_UNIX = 1602720000n; // 2020‑10‑15 00:00:00 UTC
    static ATTO_FACTOR = 1000000000000000000n; // 1e18
    static FACTOR_1E12 = 1000000000000n; // 1e12
    static V1_ACCURACY = 100000000n; // 1e8
    static V1_INFLATION_PCT_NUM = 107n;
    static V1_INFLATION_PCT_DEN = 100n;
    static PERIOD_SEC = 31556952n; // 365.25 days
    // ───────────────────────────── fixed‑point helpers ───────────────────────────
    /**
     * Multiplies two 64.64 factors and returns a 64.64 factor.
     * (a * b) >> 64
     */
    static mul64(a, b) {
        return (a * b) >> 64n;
    }
    /**
     * Multiplies a 64.64 factor with an unsigned integer (bigint) and truncates.
     * Equivalent to ABDKMath64x64.mulu.
     */
    static mulU(factor64x64, value) {
        return (factor64x64 * value) >> 64n;
    }
    /**
     * Exponentiation by squaring for 64.64 factors.
     * Equivalent to ABDKMath64x64.pow(base, exp).
     */
    static pow64(base64x64, exp) {
        let base = base64x64;
        let exponent = exp;
        let result = this.ONE_64;
        while (exponent > 0n) {
            if ((exponent & 1n) === 1n) {
                result = this.mul64(result, base);
            }
            base = this.mul64(base, base);
            exponent >>= 1n;
        }
        return result;
    }
    /** 1.0 in 1 × 10³⁶ representation. */
    static ONE_36 = 1000000000000000000000000000000000000000n; // 1e36
    /** 0.93^(1 / 365.25) scaled to 1e36 (rounded half-up). */
    static GAMMA_36 = 999801332008598957430613406568191166n;
    /** 1 / GAMMA scaled to 1e36 (rounded half-up). */
    static BETA_36 = 1000198707468214629156271489013303962n;
    /** (a · b) / 1e36 – stays inside the 1e36 domain. */
    static mul36(a, b) {
        return (a * b) / this.ONE_36;
    }
    /** Exponentiation for 1e36-scaled factors. */
    static pow36(base36, exp) {
        let result = this.ONE_36;
        let base = base36;
        let e = exp;
        while (e > 0n) {
            const isOdd = (e & 1n) === 1n;
            if (isOdd) {
                result = this.mul36(result, base);
            }
            base = this.mul36(base, base);
            e >>= 1n;
        }
        return result;
    }
    // ───────────────────────────── API: human units ─────────────────────────────
    /** Atto‑circles (1e18) → UI circles as JS number (lossless until 2^53‑1). */
    static attoCirclesToCircles(atto) {
        if (atto === 0n)
            return 0;
        // Split integer and fractional parts before casting to preserve precision.
        const whole = atto / this.ATTO_FACTOR;
        const frac = atto % this.ATTO_FACTOR;
        const MAX_SAFE_INT = BigInt(Number.MAX_SAFE_INTEGER);
        if (whole > MAX_SAFE_INT || whole < -MAX_SAFE_INT) {
            throw new RangeError('Atto value’s integer component exceeds JS double precision.');
        }
        return Number(whole) + Number(frac) / Number(this.ATTO_FACTOR);
    }
    /** UI circles → atto‑circles (truncates to match Solidity’s rounding). */
    static circlesToAttoCircles(circles) {
        return BigInt(Math.trunc(circles * Number(this.ATTO_FACTOR)));
    }
    // ───────────────────────────── API: demurrage math ──────────────────────────
    /** Inflationary → demurraged for explicit day index. */
    static inflationaryToDemurrage(inflationary, day) {
        return this.mulU(this.pow64(this.GAMMA_64, day), inflationary);
    }
    /** Demurraged → inflationary for explicit day index. */
    static demurrageToInflationary(demurraged, day) {
        return this.mulU(this.pow64(this.BETA_64, day), demurraged);
    }
    /** UNIX timestamp (seconds) → Circles day index. */
    static dayFromTimestamp(unixSeconds) {
        return (unixSeconds - this.INFLATION_DAY_ZERO_UNIX) / this.SECONDS_PER_DAY;
    }
    /** Demurraged → static circles for “today”. */
    static attoCirclesToAttoStaticCircles(demurraged, nowUnixSeconds = BigInt(Math.floor(Date.now() / 1000))) {
        return this.demurrageToInflationary(demurraged, this.dayFromTimestamp(nowUnixSeconds));
    }
    /** Static circles → demurraged circles for “today”. */
    static attoStaticCirclesToAttoCircles(staticCircles, nowUnixSeconds = BigInt(Math.floor(Date.now() / 1000))) {
        return this.inflationaryToDemurrage(staticCircles, this.dayFromTimestamp(nowUnixSeconds));
    }
    /** Inflationary → demurraged (exact, reversible). */
    static inflationaryToDemurrageExact(inflationary, day) {
        const factor = this.pow36(this.GAMMA_36, day);
        return (inflationary * factor) / this.ONE_36;
    }
    /** Demurraged → inflationary (inverse of the above, exact). */
    static demurrageToInflationaryExact(demurraged, day) {
        const factor = this.pow36(this.BETA_36, day);
        return (demurraged * factor) / this.ONE_36;
    }
    /** Demurraged atto-circles → static atto-circles “today” (loss-less). */
    static attoCirclesToAttoStaticCirclesExact(demurraged, nowUnixSeconds = BigInt(Math.floor(Date.now() / 1000))) {
        const day = this.dayFromTimestamp(nowUnixSeconds);
        return this.demurrageToInflationaryExact(demurraged, day);
    }
    /** Static atto-circles → demurraged atto-circles “today” (loss-less). */
    static attoStaticCirclesToAttoCirclesExact(staticCircles, nowUnixSeconds = BigInt(Math.floor(Date.now() / 1000))) {
        const day = this.dayFromTimestamp(nowUnixSeconds);
        return this.inflationaryToDemurrageExact(staticCircles, day);
    }
    // ───────────────────── utilities for 6‑decimal truncation ───────────────────
    static truncateToInt64(wei) {
        const truncated = wei / this.FACTOR_1E12;
        const MAX_INT64 = 9223372036854775807n;
        return truncated > MAX_INT64 ? MAX_INT64 : truncated;
    }
    static blowUpToBigInt(sixDecimals) {
        return sixDecimals * this.FACTOR_1E12;
    }
    static truncateToSixDecimals(wei) {
        return this.blowUpToBigInt(this.truncateToInt64(wei));
    }
    // ───────────────────────── v1 → v2 migration helpers ─────────────────────────
    static v1InflateFactor(periodIdx) {
        if (periodIdx === 0n)
            return this.V1_ACCURACY;
        return (this.V1_ACCURACY * this.V1_INFLATION_PCT_NUM ** periodIdx) / (this.V1_INFLATION_PCT_DEN ** periodIdx);
    }
    /** CRC → demurraged Circles for a given timestamp. */
    static attoCrcToAttoCircles(v1Amount, blockTimestampUtc) {
        const secondsSinceEpoch = blockTimestampUtc - this.INFLATION_DAY_ZERO_UNIX;
        const periodIdx = secondsSinceEpoch / this.PERIOD_SEC;
        const secondsIntoPeriod = secondsSinceEpoch % this.PERIOD_SEC;
        const factorCur = this.v1InflateFactor(periodIdx);
        const factorNext = this.v1InflateFactor(periodIdx + 1n);
        return this.v1ToDemurrage(v1Amount, factorCur, factorNext, secondsIntoPeriod, this.PERIOD_SEC);
    }
    /** Demurraged Circles → CRC (inverse of the above). */
    static attoCirclesToAttoCrc(demurraged, blockTimestampUtc) {
        const secondsSinceEpoch = blockTimestampUtc - this.INFLATION_DAY_ZERO_UNIX;
        const periodIdx = secondsSinceEpoch / this.PERIOD_SEC;
        const secondsIntoPeriod = secondsSinceEpoch % this.PERIOD_SEC;
        const factorCur = this.v1InflateFactor(periodIdx);
        const factorNext = this.v1InflateFactor(periodIdx + 1n);
        const rP = factorCur * (this.PERIOD_SEC - secondsIntoPeriod) + factorNext * secondsIntoPeriod;
        return (demurraged * 3n * this.V1_ACCURACY * this.PERIOD_SEC) / rP;
    }
    /** Implements the on‑chain formula used during migration. */
    static v1ToDemurrage(v1Amount, factorCur, factorNext, secondsInto, periodSec) {
        const rP = factorCur * (periodSec - secondsInto) + factorNext * secondsInto;
        return (v1Amount * 3n * this.V1_ACCURACY * periodSec) / rP;
    }
}

/**
 * Converts a CIDv0 string to a UInt8Array, stripping the hashing algorithm identifier.
 * @param {string} cidV0 - The CIDv0 string (e.g., Qm...).
 * @returns {Uint8Array} - The resulting UInt8Array of the 32-byte hash digest.
 */
function cidV0ToUint8Array(cidV0) {
    // Decode the base58 CIDv0 string to a Multihash
    const multihashBytes = multihash.fromB58String(cidV0);
    // Verify the multihash algorithm (should be SHA-256)
    const decodedMultihash = multihash.decode(multihashBytes);
    if (decodedMultihash.code !== multihash.names['sha2-256']) {
        throw new Error('Unsupported hash algorithm. Only SHA-256 is supported for CIDv0.');
    }
    // Extract and return the 32-byte hash digest
    return decodedMultihash.digest;
}
/**
 * Converts a 32-byte UInt8Array back to a CIDv0 string by adding the hashing algorithm identifier.
 * @param {Uint8Array} uint8Array - The 32-byte hash digest.
 * @returns {string} - The resulting CIDv0 string (e.g., Qm...).
 */
function uint8ArrayToCidV0(uint8Array) {
    if (uint8Array.length !== 32) {
        throw new Error('Invalid array length. Expected 32 bytes.');
    }
    // Recreate the Multihash (prefix with SHA-256 code and length)
    const multihashBytes = multihash.encode(uint8Array, 'sha2-256');
    // Encode the Multihash as a base58 CIDv0 string
    return multihash.toB58String(multihashBytes);
}
/**
 * Converts a Uint8Array to a hex string.
 * @param uint8Array - The Uint8Array to convert.
 */
function uint8ArrayToHexString(uint8Array) {
    return Array.from(uint8Array).map(byte => byte.toString(16).padStart(2, '0')).join('');
}
/**
 * Converts a hex string to a Uint8Array.
 * @param {string} hexString - The hex string to convert.
 * @returns {Uint8Array} - The resulting Uint8Array.
 */
function hexStringToUint8Array(hexString) {
    const bytes = [];
    for (let i = 0; i < hexString.length; i += 2) {
        bytes.push(parseInt(hexString.substr(i, 2), 16));
    }
    return new Uint8Array(bytes);
}
/**
 * Converts a string to a Uint8Array.
 * @param {string} str - The string to convert.
 * @returns {Uint8Array} - The resulting Uint8Array.
 */
function stringToUint8Array(str) {
    const encoder = new TextEncoder();
    return encoder.encode(str);
}
function addressToUInt256(address) {
    // Remove the '0x' prefix if it exists
    let addr = address;
    if (addr.startsWith('0x')) {
        addr = addr.slice(2);
    }
    // Convert the address to a BigInt
    return BigInt('0x' + addr);
}
function uint256ToAddress(uint256) {
    // Convert the BigInt to a hex string
    let hexString = uint256.toString(16);
    // Add leading zeros to ensure the string length is 40
    hexString = hexString.padStart(40, '0');
    // Add the '0x' prefix
    return `0x${hexString}`;
}
const errorAbis = [
    'error Create2InsufficientBalance(uint256 balance, uint256 needed)',
    'error Create2EmptyBytecode()',
    'error Create2FailedDeployment()',
    'error ReentrancyGuardReentrantCall()',
    'error AddressInsufficientBalance(address account)',
    'error AddressEmptyCode(address target)',
    'error FailedInnerCall()',
    'error ERC20InsufficientBalance(address sender, uint256 balance, uint256 needed)',
    'error ERC20InvalidSender(address sender)',
    'error ERC20InvalidReceiver(address receiver)',
    'error ERC20InsufficientAllowance(address spender, uint256 allowance, uint256 needed)',
    'error ERC20InvalidApprover(address approver)',
    'error ERC20InvalidSpender(address spender)',
    'error ERC721InvalidOwner(address owner)',
    'error ERC721NonexistentToken(uint256 tokenId)',
    'error ERC721IncorrectOwner(address sender, uint256 tokenId, address owner)',
    'error ERC721InvalidSender(address sender)',
    'error ERC721InvalidReceiver(address receiver)',
    'error ERC721InsufficientApproval(address operator, uint256 tokenId)',
    'error ERC721InvalidApprover(address approver)',
    'error ERC721InvalidOperator(address operator)',
    'error ERC1155InsufficientBalance(address sender, uint256 balance, uint256 needed, uint256 tokenId)',
    'error ERC1155InvalidSender(address sender)',
    'error ERC1155InvalidReceiver(address receiver)',
    'error ERC1155MissingApprovalForAll(address operator, address owner)',
    'error ERC1155InvalidApprover(address approver)',
    'error ERC1155InvalidOperator(address operator)',
    'error ERC1155InvalidArrayLength(uint256 idsLength, uint256 valuesLength)',
    'error OwnableUnauthorizedAccount(address account)',
    'error OwnableInvalidOwner(address owner)',
    'error CirclesErrorNoArgs(uint8)',
    'error CirclesErrorOneAddressArg(address, uint8)',
    'error CirclesErrorAddressUintArgs(address, uint256, uint8)',
    'error CirclesHubFlowEdgeStreamMismatch(uint256 flowEdgeId, uint256 streamId, uint8 code)',
    'error CirclesHubStreamMismatch(uint256 streamId)',
    'error CirclesHubNettedFlowMismatch(uint256 vertexPosition, int256 matrixNettedFlow, int256 streamNettedFlow)',
    'error CirclesERC1155CannotReceiveBatch(uint8 code)',
    'error CirclesInvalidCirclesId(uint256 id, uint8 code)',
    'error CirclesInvalidParameter(uint256 parameter, uint8 code)',
    'error CirclesAmountOverflow(uint256 amount, uint8 code)',
    'error CirclesProxyAlreadyInitialized()',
    'error CirclesIdMustBeDerivedFromAddress(uint256 providedId, uint8 code)',
    'error CirclesReentrancyGuard(uint8 code)',
    'error CirclesStandardTreasuryGroupHasNoVault(address group)',
    'error CirclesStandardTreasuryInvalidMetadataType(bytes32 metadataType, uint8 code)',
    'error CirclesStandardTreasuryInvalidMetadata(bytes metadata, uint8 code)',
    'error CirclesNamesInvalidName(address avatar, string name, uint8 code)',
    'error CirclesNamesShortNameAlreadyAssigned(address avatar, uint72 shortName, uint8 code)',
    'error CirclesNamesShortNameWithNonceTaken(address avatar, uint256 nonce, uint72 shortName, address takenByAvatar)',
    'error CirclesNamesAvatarAlreadyHasCustomNameOrSymbol(address avatar, string nameOrSymbol, uint8 code)',
    'error CirclesNamesOrganizationHasNoSymbol(address organization, uint8 code)',
    'error CirclesNamesShortNameZero(address avatar, uint256 nonce)',
    'error CirclesMigrationAmountMustBeGreaterThanZero()',
    'error InflationaryCirclesOperatorOnlyActOnBalancesOfSender(address sender, address from)',
    'error CirclesOperatorInvalidStreamSource(uint256 streamIndex, uint256 singleSourceCoordinate, uint256 streamSourceCoordinate)',
    'error CirclesStandardTreasuryRedemptionCollateralMismatch(uint256 circlesId, uint256[] redemptionIds, uint256[] redemptionValues, uint256[] burnIds, uint256[] burnValues)',
    'error ERC2612ExpiredSignature(uint256 deadline)',
    'error ERC2612InvalidSigner(address signer, address owner)',
    // Circles Group operator contracts errors
    'error SupergroupProxyAlreadyInitialised()',
    'error SupergroupOnlyHub()',
    'error SupergroupOnlyOwner()',
    'error SupergroupOnlyOwnerOrService()',
    'error SupergroupOnlyAuthorizedOperator()',
    'error SupergroupMustBeRegistered()',
    'error SupergroupMustUseStandardTreasury()',
    'error SupergroupInvalidCallingParameters()',
    'error SupergroupBlockNormalERC1155Transfers()',
    'error SupergroupAlwaysBlockUntrustedIds()',
    'error SupergroupInvalidOperator(address operator)',
    'error SupergroupMustHaveOperatorsActivated()',
    'error SupergroupLogicAssertion()',
    'error SupergroupOperatorRequestInProgress()',
    'error SupergroupFingerprintUnderflow()',
    'error SupergroupOperatorUnservicedGroup(address group)',
    'error SupergroupOperatorNotAuthorizedAndAuthorizationRequired(address group)',
    'error SupergroupOperatorDoesNotImplement()',
    'error ExpectationAlreadySet(bytes32 expectation)',
    'error NoExpectationSet()',
    'error ExpectationMismatch(bytes32 expected, bytes32 actual)',
    'error ExpectationSingleReceiveOnlySupergroupId(uint256 id)',
    'error CMGroupProxyAlreadyInitialised()',
    'error CMGroupOnlyHub()',
    'error CMGroupOnlyHubOrTreasury()',
    'error CMGroupOnlyOwner()',
    'error CMGroupMembershipCheckFailed(address avatar, address failedCondition)',
    'error CMGroupMaxConditionsActive(uint256 conditionsActive)',
    'error CMGroupOnlyOwnerOrService()',
    'error CMGroupInteractionAmountIsBelowMinimum(uint256 id, uint256 receivedAmount, uint256 minimalAmount)',
    'error CMGroupInvalidCallingParameters()',
    'error CMGHandlerOnlyHub()',
    'error CMGHandlerOnlyCMGroup()',
    'error CMGHandlerOnlyOwner()',
    'error CMGHandlerAcceptanceCallUnhandled()',
    'error CMGHandlerOperatorUnservicedGroup(address group)',
    'error CMGHandlerRefuseGroupCircles()',
    'error CMGHandlerConversionOngoing(uint256 amount)',
    'error CMGHandlerNoConversionExpected()',
    'error CMGHandlerReceivedZeroAmount()',
    'error CMGHandlerRedemptionExpectedFromVault(address from)',
    'error CMGHandlerDataHashMismatchUponReceiving(bytes32 expectedDataHash, bytes receivedData)',
    'error CMGHandlerVaultNotFound(address group)',
    'error CMGHandlerCouldNotFillRedemptionRequest()',
    'error CMGHandlerEarlyRevertCollateralNotPresent()',
    'error CMGHandlerOnlyTransferOwnCircles()',
    'error CMGHandlerInvalidCallingParameters()',
    'error CMGHandlerLogicAssertion()',
    'error CMGPrimaryGroupMustBeHumanAndGroupToRegisterPrimaryGroup(address human, address group)',
    'error CMGRedemptionOperatorGroupMustBeRegistered(address group)',
    'error CMGRedemptionOperatorHandlerOfGroupZeroAddress()',
    'error CMGRedemptionOperatorOnlySupportStandardTreasuryGroups(address treasury)',
    'error CMGRedemptionOperatorFoundCollateralExceedsAmountRequested(uint256 requestedAmount, uint256 foundAmount)',
    'error CMGRedemptionOperatorFailedToFindSufficientCollateral(uint256 requestedAmount, uint256 foundAmount)',
    'error CMGRedemptionOperatorInvalidCallingParameters()',
    'error GroupLiquidityProviderGroupCannotBeZeroAddress()',
    'error GroupLiquidityProviderOwnerCannotBeZeroAddress()',
    'error GroupLiquidityProviderCanOnlyTransferOwnTokens()',
    'error GroupLiquidityProviderOnlyAcceptTransfersFromHub()',
    'error GroupLiquidityProviderOnlyAcceptTransfersFromOwnerOrVault()',
    // Liquidity bootstrapping pool factory errors
    'error OnlyFactory()',
    'error OrderNotFilledYet()',
    'error AlreadyCreated()',
    'error InsufficientBackingAssetBalance()',
    'error NotBacker()',
    'error TokensLockedUntilTimestamp(uint256 timestamp)',
    'error OnlyHub()',
    'error NotExactlyRequiredCRCAmount(uint256 required, uint256 received)',
    'error OnlyHumanAvatarsAreSupported()',
    'error BackingInFavorDissalowed()',
    'error UnsupportedBackingAsset(address requestedAsset)',
    'error CirclesBackingDeploymentFailed(address backer)',
    'error OnlyCirclesBacking()',
    'error NotAdmin()',
    'error OnlyTwoTokenLBPSupported()'
];
const errorInterface = new ethers.Interface(errorAbis);
// Parsing the error data using the updated ABI definitions
function parseError(errorData) {
    try {
        return errorInterface.parseError(errorData);
    }
    catch (error) {
        throw new Error(`Error decoding the revert data: ${error}. Original error data: ${errorData}`);
    }
}
function getCustomErrorFragment(errorText) {
    const abi = [
        'error Error(string)'
    ];
    const iface = new ethers.Interface(abi);
    const errorFragment = iface.getError('Error');
    const encodedError = iface.encodeErrorResult('Error', [errorText]);
    const decodedError = iface.decodeErrorResult('Error', encodedError);
    return {
        fragment: errorFragment,
        name: errorFragment?.name,
        args: decodedError,
        signature: errorFragment?.format(),
        selector: iface.getFunction('Error')?.selector
    };
}
function handleTransactionError(e) {
    console.log(e);
    // handle when data is null
    if (e.data) {
        const parsedError = parseError(e.data);
        if (parsedError) {
            const bigIntReplacer = (key, value) => {
                if (typeof value === 'bigint') {
                    return value.toString();
                }
                return value;
            };
            throw new Error(JSON.stringify(parsedError, bigIntReplacer, 2));
        }
    }
    else if (!e.data && e.info?.error?.message) {
        throw new Error(JSON.stringify(getCustomErrorFragment(e.info.error.message), null, 2));
    }
    throw new Error(JSON.stringify(getCustomErrorFragment('Unknown'), null, 2));
}
// Utility function to generate mint data based on type
function generateBaseGroupMintData(mintType) {
    switch (mintType) {
        case 'ERC1155':
            return new Uint8Array(0);
        case 'TYPE_INFLATIONARY':
            const hashTypeInflationary = ethers.keccak256(ethers.toUtf8Bytes("TYPE_INFLATIONARY"));
            return ethers.getBytes(hashTypeInflationary);
        case 'TYPE_DEMURRAGE':
            const hashDemurage = ethers.keccak256(ethers.toUtf8Bytes("TYPE_DEMURRAGE"));
            return ethers.getBytes(hashDemurage);
        default:
            throw new Error(`Unsupported token type: ${mintType}`);
    }
}

export { BaseGroupMintType, CirclesConverter, addressToUInt256, cidV0ToUint8Array, errorAbis, generateBaseGroupMintData, handleTransactionError, hexStringToUint8Array, parseError, stringToUint8Array, uint256ToAddress, uint8ArrayToCidV0, uint8ArrayToHexString };
