// @circles-sdk/sdk/dist/index.js
import { ethers, keccak256, toUtf8Bytes, formatEther, ZeroAddress, AbiCoder } from 'ethers';
import { Token__factory, Hub__factory, NameRegistryV1__factory } from '@circles-sdk/abi-v1';
import { cidV0ToUint8Array, CirclesConverter, addressToUInt256 } from '@circles-sdk/utils';
import { BaseGroup__factory, CoreMembersGroup__factory, Hub__factory as Hub__factory$1, NameRegistry__factory, CMGroupDeployer__factory, BaseGroupFactory__factory, Migration__factory, InflationaryCircles__factory, DemurrageCircles__factory } from '@circles-sdk/abi-v2';
import { findMaxFlow, findPath, getTokenInfoMapFromPath, createFlowMatrix } from '@circles-sdk/pathfinder';
import { CirclesRpc, CirclesData } from '@circles-sdk/data';
export { Observable } from '@circles-sdk/data';
import { Profiles } from '@circles-sdk/profiles';

class V1Avatar {
    sdk;
    get address() {
        if (!this.avatarInfo) {
            throw new Error('Avatar is not initialized');
        }
        return this.avatarInfo.avatar;
    }
    get v1Token() {
        return this._v1Token;
    }
    _v1Token;
    avatarInfo;
    constructor(sdk, avatarInfo) {
        this.sdk = sdk;
        this.avatarInfo = avatarInfo;
        if (!this.avatarInfo.hasV1) {
            throw new Error('Avatar is not a v1 avatar');
        }
        if (this.avatarInfo.v1Token) {
            this._v1Token = Token__factory.connect(this.avatarInfo.v1Token, this.sdk.contractRunner);
        }
    }
    async getBalances() {
        return await this.sdk.data.getTokenBalances(this.address);
    }
    async updateMetadata(cid) {
        this.throwIfNotInitialized();
        const digest = cidV0ToUint8Array(cid);
        const tx = await this.sdk.v1NameRegistry?.updateMetadataDigest(digest);
        const receipt = await tx?.wait();
        if (!receipt) {
            throw new Error('Metadata update failed');
        }
        this.avatarInfo.cidV0 = cid;
        return receipt;
    }
    /**
     * Utilizes the pathfinder to find the max. transferable amount from the avatar to `to`.
     * @param to The recipient
     * @param tokenId The token to transfer (address). Leave empty to allow transitive transfers.
     * @returns The max. transferable amount at the time.
     */
    async getMaxTransferableAmount(to, tokenId) {
        this.throwIfNotInitialized();
        if (tokenId) {
            const tokenInfo = await this.sdk.data.getTokenInfo(tokenId);
            if (!tokenInfo) {
                throw new Error('Token not found');
            }
            const tokenBalances = await this.sdk.data.getTokenBalances(this.address);
            const tokenBalance = tokenBalances.filter(b => b.version === 1 && b.tokenAddress === tokenId)[0]?.circles;
            return tokenBalance ?? '0';
        }
        this.throwIfPathfinderIsNotAvailable();
        const largeAmount = BigInt('999999999999999999999999999999');
        const transferPath = await this.sdk.v1Pathfinder.getPath(this.address, to, largeAmount.toString());
        if (!transferPath.transfers.length) {
            return 0;
        }
        const attoCircles = CirclesConverter.attoCrcToAttoCircles(BigInt(transferPath.maxFlow), BigInt(Math.floor(Date.now() / 1000)));
        return CirclesConverter.attoCirclesToCircles(attoCircles);
    }
    /**
     * Utilizes the pathfinder to transitively send `amount` Circles to `to`.
     * @param to The recipient
     * @param amount The amount to send
     * @param token The token to transfer (address). Leave empty to allow transitive transfers.
     */
    async transfer(to, amount, token) {
        this.throwIfNotInitialized();
        let receipt = null;
        if (!token) {
            this.throwIfPathfinderIsNotAvailable();
            // transitive transfer
            const transferPath = await this.sdk.v1Pathfinder.getPath(this.address, to, amount.toString());
            if (transferPath.transfers.length === 0) {
                throw new Error(`Couldn't find a valid path from ${this.address} to ${to} for ${amount}.`);
            }
            const tokenOwners = transferPath.transfers.map(o => o.tokenOwner);
            const srcs = transferPath.transfers.map(o => o.from);
            const dests = transferPath.transfers.map(o => o.to);
            const wads = transferPath.transfers.map(o => BigInt(o.value));
            const tx = await this.sdk.v1Hub.transferThrough(tokenOwners, srcs, dests, wads);
            receipt = await tx.wait();
        }
        else {
            // erc20 transfer via ethers.Interface
            const iface = new ethers.Interface(['function transfer(address to, uint256 value)']);
            const data = iface.encodeFunctionData('transfer', [to, amount]);
            if (!this.sdk?.contractRunner?.sendTransaction) {
                throw new Error('ContractRunner not available');
            }
            const tx = await this.sdk.contractRunner.sendTransaction({
                to: token,
                data: data,
                value: 0n
            });
            receipt = tx;
        }
        if (!receipt) {
            throw new Error(`The transferThrough call for '${this.address} -> ${to}: ${amount}' didn't yield a receipt.`);
        }
        return receipt;
    }
    async trust(avatar) {
        this.throwIfNotInitialized();
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        const avatars = Array.isArray(avatar) ? avatar : [avatar];
        const batch = this.sdk.contractRunner.sendBatchTransaction();
        for (const av of avatars) {
            const txData = this.sdk.v1Hub.interface.encodeFunctionData('trust', [av, BigInt(100)]);
            batch.addTransaction({
                to: this.sdk.circlesConfig.v1HubAddress,
                data: txData,
                value: 0n
            });
        }
        const receipt = await batch.run();
        if (!receipt) {
            throw new Error('Trust failed');
        }
        return receipt;
    }
    async untrust(avatar) {
        this.throwIfNotInitialized();
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        const avatars = Array.isArray(avatar) ? avatar : [avatar];
        const batch = this.sdk.contractRunner.sendBatchTransaction();
        for (const av of avatars) {
            const txData = this.sdk.v1Hub.interface.encodeFunctionData('trust', [av, BigInt(0)]);
            batch.addTransaction({
                to: this.sdk.circlesConfig.v1HubAddress,
                data: txData,
                value: 0n
            });
        }
        const receipt = await batch.run();
        if (!receipt) {
            throw new Error('Untrust failed');
        }
        return receipt;
    }
    async getMintableAmount() {
        if (!this.v1Token) {
            return 0;
        }
        const availableCrcToMint = await this.v1Token.look();
        const attoCircles = CirclesConverter.attoCrcToAttoCircles(availableCrcToMint, BigInt(Math.floor(Date.now() / 1000)));
        const mintableCircles = CirclesConverter.attoCirclesToCircles(attoCircles);
        return mintableCircles;
    }
    async personalMint() {
        this.throwIfNotInitialized();
        if (!this.v1Token) {
            throw new Error('Avatar does not have a token to mint');
        }
        if (await this.v1Token.stopped()) {
            throw new Error('Avatar token is stopped');
        }
        const tx = await this.v1Token.update();
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('The update call didn\'t yield a receipt');
        }
        return receipt;
    }
    async stop() {
        this.throwIfNotInitialized();
        if (!this.v1Token) {
            throw new Error('Avatar does not have a token to stop');
        }
        if (await this.v1Token.stopped()) {
            throw new Error('Avatar token is already stopped');
        }
        const tx = await this.v1Token.stop();
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('The stop call didn\'t yield a receipt');
        }
        return receipt;
    }
    throwIfNotInitialized() {
        if (this.avatarInfo) {
            return;
        }
        throw new Error('Avatar is either not initialized or is not signed up at Circles.');
    }
    async getTrustRelations() {
        return this.sdk.data.getAggregatedTrustRelations(this.address, 1);
    }
    async getTransactionHistory(pageSize) {
        const query = this.sdk.data.getTransactionHistory(this.address, pageSize);
        await query.queryNextPage();
        return query;
    }
    async getTotalBalance() {
        return parseFloat(await this.sdk.data.getTotalBalance(this.address, true));
    }
    async getGasTokenBalance() {
        // TODO: re-implement
        // return await this.sdk.contractRunner.provider?.getBalance(this.address) ?? 0n;
        return 0n;
    }
    async trusts(otherAvatar) {
        return (await this.sdk.v1Hub.limits(this.address, otherAvatar)) > 0n;
    }
    async isTrustedBy(otherAvatar) {
        return (await this.sdk.v1Hub.limits(otherAvatar, this.address)) > 0n;
    }
    /**
     * Gets the total supply of either this avatar's Personal Circles.
     * Returns '0' for organizations or if the avatar is not signed up at Circles.
     */
    async getTotalSupply() {
        this.throwIfNotInitialized();
        return await this._v1Token?.totalSupply() ?? 0n;
    }
    throwIfPathfinderIsNotAvailable() {
        if (!this.sdk.v1Pathfinder) {
            throw new Error('Pathfinder is not available');
        }
    }
}

class V2Avatar {
    sdk;
    get address() {
        return this.avatarInfo.avatar;
    }
    avatarInfo;
    _cachedProfile;
    _cachedProfileCid;
    METADATATYPE_GROUPREDEEM = keccak256(toUtf8Bytes('CIRCLESv2:RESERVED_DATA:CirclesGroupRedeem'));
    constructor(sdk, avatarInfo) {
        this.sdk = sdk;
        this.avatarInfo = avatarInfo;
        if (this.avatarInfo.version != 2) {
            throw new Error('Avatar is not a v2 avatar');
        }
    }
    trusts(otherAvatar) {
        return this.sdk.v2Hub.isTrusted(this.address, otherAvatar);
    }
    isTrustedBy(otherAvatar) {
        return this.sdk.v2Hub.isTrusted(otherAvatar, this.address);
    }
    async updateMetadata(cid) {
        this.throwIfNameRegistryIsNotAvailable();
        const digest = cidV0ToUint8Array(cid);
        const tx = await this.sdk.nameRegistry?.updateMetadataDigest(digest);
        const receipt = await tx?.wait();
        if (!receipt) {
            throw new Error('Transfer failed');
        }
        this.avatarInfo.cidV0 = cid;
        return receipt;
    }
    async getMaxTransferableAmount(to, tokenId, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers) {
        this.throwIfV2IsNotAvailable();
        to = to.toLowerCase();
        excludeFromTokens = await this.sdk.getDefaultTokenExcludeList(to, excludeFromTokens);
        if (tokenId) {
            const tokenInfo = await this.sdk.data.getTokenInfo(tokenId);
            if (!tokenInfo) {
                throw new Error('Token not found');
            }
            const tokenBalances = await this.sdk.data.getTokenBalances(this.address);
            const tokenBalance = tokenBalances.filter(b => b.version === 2 && b.tokenAddress === tokenInfo.token)[0];
            return tokenBalance?.circles ?? 0;
        }
        const result = await findMaxFlow(this.sdk.circlesConfig.circlesRpcUrl, {
            from: this.address,
            to,
            useWrappedBalances,
            fromTokens,
            toTokens,
            excludeFromTokens,
            excludeToTokens,
            maxTransfers
        });
        return CirclesConverter.attoCirclesToCircles(CirclesConverter.truncateToSixDecimals(result));
    }
    async getMintableAmount() {
        this.throwIfV2IsNotAvailable();
        const [a, _, __] = await this.sdk.v2Hub.calculateIssuance(this.address);
        return parseFloat(formatEther(a));
    }
    async getTotalBalance() {
        return parseFloat(await this.sdk.data.getTotalBalanceV2(this.address, true));
    }
    async getGasTokenBalance() {
        // TODO: re-implement
        // return await this.sdk.contractRunner.provider?.getBalance(this.address) ?? 0n;
        return 0n;
    }
    async getTransactionHistory(pageSize) {
        const query = this.sdk.data.getTransactionHistory(this.address, pageSize);
        await query.queryNextPage();
        return query;
    }
    async getTrustRelations() {
        return this.sdk.data.getAggregatedTrustRelations(this.address, 2);
    }
    async getBalances() {
        return await this.sdk.data.getTokenBalances(this.address);
    }
    async personalMint() {
        this.throwIfV2IsNotAvailable();
        const tx = await this.sdk.v2Hub.personalMint();
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Personal mint failed');
        }
        return receipt;
    }
    async stop() {
        this.throwIfV2IsNotAvailable();
        const tx = await this.sdk.v2Hub.stop();
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Stop failed');
        }
        return receipt;
    }
    async transitiveTransfer(to, amount, batch, txData, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers) {
        this.throwIfV2IsNotAvailable();
        to = to.toLowerCase();
        excludeFromTokens = await this.sdk.getDefaultTokenExcludeList(to, excludeFromTokens);
        amount = CirclesConverter.truncateToSixDecimals(amount);
        /*
          1) Ask the pathfinder for a transfer path.
    
          2) Collect all wrapped-token edges that originate at the **sender**:
             a) static wrappers (type = CrcV2_ERC20WrapperDeployed_Inflationary)
             b) demurraged wrappers (type = CrcV2_ERC20WrapperDeployed_Demurraged)
    
          3) For every *static* wrapper found in 2 a), fetch the sender’s **entire**
             static balance (in atto-static circles).
    
          4) Build `unwrap()` calls
               – static wrappers: unwrap *full* balance
               – demurraged wrappers: unwrap **exact** amount needed by the path
             → after this step the Safe holds plain atto-Circles for every wrapper
               token that will be moved.
    
          5) Rewrite the path so that each wrapped edge now references the real
             token owner (`tokenOwner = underlying ERC-1155 ID`), and
             – keep a running total of how many demurraged units of every static
               wrapper will actually be spent.
    
          6) For each static wrapper:
               remaining = (unwrapped total) − (spent in step 5)
             Add a `Hub.wrap(realOwner, remaining, 1)` call so the leftover
             Circles are re-wrapped to static ERC-20 after the transfer.
             (Demurraged wrappers have no leftovers by design.)
    
          7) Build the `FlowMatrix` from the rewritten transfers.
    
          8) Assemble the Safe MultiSend in *strict* order
               self-approval → all unwraps → operateFlowMatrix → all re-wraps
    
          9) Ask the Safe for a gas estimate; abort if it returns 0.
        */
        const path = await findPath(this.sdk.circlesConfig.circlesRpcUrl, {
            from: this.address,
            to,
            targetFlow: amount.toString(),
            useWrappedBalances,
            fromTokens,
            toTokens,
            excludeFromTokens,
            excludeToTokens,
            maxTransfers
        });
        const wrapCalls = [];
        const unwrapCalls = [];
        // Get the token info for all tokens in the path
        const tokenInfoMap = await getTokenInfoMapFromPath(this.sdk.circlesConfig.circlesRpcUrl, path);
        // Find all wrapped edges (can only originate from the sender)
        const allWrappedEdges = path.transfers
            .filter(o => o.from == this.avatarInfo.avatar)
            .filter(o => !!tokenInfoMap.get(o.tokenOwner.toLowerCase())?.type.startsWith('CrcV2_ERC20WrapperDeployed'));
        // Filter the static edges
        const wrappedStaticEdges = allWrappedEdges.filter(o => tokenInfoMap.get(o.tokenOwner.toLowerCase())?.type === 'CrcV2_ERC20WrapperDeployed_Inflationary');
        const wrapedStaticEdgeTotalsByToken = {};
        wrappedStaticEdges.forEach(o => {
            if (!wrapedStaticEdgeTotalsByToken[o.tokenOwner]) {
                wrapedStaticEdgeTotalsByToken[o.tokenOwner] = BigInt(0);
            }
            wrapedStaticEdgeTotalsByToken[o.tokenOwner] += BigInt(o.value);
        });
        // Filter the demurraged edges
        const wrappedDemurragedEdges = allWrappedEdges.filter(o => tokenInfoMap.get(o.tokenOwner.toLowerCase())?.type === 'CrcV2_ERC20WrapperDeployed_Demurraged');
        const wrappedDemurragedEdgeTotalsByToken = {};
        wrappedDemurragedEdges.forEach(o => {
            if (!wrappedDemurragedEdgeTotalsByToken[o.tokenOwner]) {
                wrappedDemurragedEdgeTotalsByToken[o.tokenOwner] = BigInt(0);
            }
            wrappedDemurragedEdgeTotalsByToken[o.tokenOwner] += BigInt(o.value);
        });
        const WRAPPER_ERC20_TOKEN_ABI = [
            'function unwrap(uint256 _amount)'
        ];
        // Unwrap all used static wrapped tokens fully
        const usedStaticTokenCount = Object.keys(wrapedStaticEdgeTotalsByToken).length;
        const wrappedStaticBalanceByTokenInStaticUnits = {};
        if (usedStaticTokenCount > 0) {
            const senderWrappedStaticTotals = await this.getStaticWrappedTokenTotalsFromSender(this.address);
            const relevantWrappedStaticBalances = senderWrappedStaticTotals.filter(o => !!wrapedStaticEdgeTotalsByToken[o.tokenAddress]);
            relevantWrappedStaticBalances.forEach(o => {
                wrappedStaticBalanceByTokenInStaticUnits[o.tokenAddress] = BigInt(o.staticAttoCircles);
                unwrapCalls.push({
                    to: o.tokenAddress,
                    data: new ethers.Interface(WRAPPER_ERC20_TOKEN_ABI).encodeFunctionData('unwrap', [BigInt(o.staticAttoCircles)]),
                    value: 0n
                });
            });
        }
        // Unwrap all used demurraged wrapped tokens exactly
        const usedDemurragedTokenCount = Object.keys(wrappedDemurragedEdgeTotalsByToken).length;
        if (usedDemurragedTokenCount > 0) {
            Object.entries(wrappedDemurragedEdgeTotalsByToken).forEach(([wrapperAddr, total]) => {
                unwrapCalls.push({
                    to: wrapperAddr,
                    data: new ethers.Interface(WRAPPER_ERC20_TOKEN_ABI).encodeFunctionData('unwrap', [total]),
                    value: 0n
                });
            });
        }
        // From this point on, we have enough (unwrapped, demurraged) tokens of each kind to
        // facilitate the transfer. However, the returned path still contains the wrapped token
        // addresses as "tokenOwner". We need to look up the real token owner and replace it
        // in the path (we create a copy).
        const unwrappedStaticTokensUsedInDemurragedUnits = {};
        const unwrappedTransfers = path.transfers.map(o => {
            const tokenInfo = tokenInfoMap.get(o.tokenOwner.toLowerCase());
            if (tokenInfo && tokenInfo.type.startsWith('CrcV2_ERC20WrapperDeployed')) {
                // Use the opportunity to also do some bookkeeping about how many demurraged tokens
                // from the unwrapped static tokens have been used so far. We need this to later know
                // how many demurraged tokens we have left to wrap them again.
                if (tokenInfo.type === 'CrcV2_ERC20WrapperDeployed_Inflationary') {
                    if (!unwrappedStaticTokensUsedInDemurragedUnits[tokenInfo.token]) {
                        unwrappedStaticTokensUsedInDemurragedUnits[tokenInfo.token] = BigInt(0);
                    }
                    unwrappedStaticTokensUsedInDemurragedUnits[tokenInfo.token] += BigInt(o.value);
                }
                return {
                    ...o,
                    tokenOwner: tokenInfo.tokenOwner
                };
            }
            return o;
        });
        // Calculate what's left after the transfer for each static token and prepare the wrap calls
        await Promise.all(Object.entries(unwrappedStaticTokensUsedInDemurragedUnits).map(async ([token, totalUsedDemurraged]) => {
            const totalStaticTokenBalance = wrappedStaticBalanceByTokenInStaticUnits[token];
            const totalDemurragedTokenBalance = CirclesConverter.attoStaticCirclesToAttoCircles(totalStaticTokenBalance);
            const remainingDemurragedBalance = totalDemurragedTokenBalance - totalUsedDemurraged;
            // Add the wrap call for the remaining static balance
            const realTokenOwner = tokenInfoMap.get(token.toLowerCase())?.tokenOwner;
            if (!realTokenOwner) {
                throw new Error(`Token owner not found for token: ${token}`);
            }
            const wrapTx = await this.sdk.v2Hub.wrap.populateTransaction(realTokenOwner, remainingDemurragedBalance, 1);
            wrapCalls.push({
                to: wrapTx.to,
                data: wrapTx.data,
                value: 0n
            });
        }));
        // Finally, we can create the flow matrix with the unwrapped transfers
        const flowMatrix = createFlowMatrix(this.address, to, path.maxFlow, unwrappedTransfers);
        // If we have data, attach it to the streams
        if (txData) {
            for (let i = 0; i < flowMatrix.streams.length; i++) {
                flowMatrix.streams[i].data = txData || new Uint8Array(0);
            }
        }
        // Create the operateFlowMatrix call
        const operateFlowMatrixCall = await this.sdk.v2Hub.operateFlowMatrix.populateTransaction(flowMatrix.flowVertices, flowMatrix.flowEdges, flowMatrix.streams, flowMatrix.packedCoordinates);
        const selfApprovalCall = await this.sdk.v2Hub.setApprovalForAll.populateTransaction(this.address, true);
        if (!selfApprovalCall) {
            throw new Error('Failed to create self-approval call');
        }
        // Prepare the Safe multi-send call
        const subCalls = [
            selfApprovalCall,
            ...unwrapCalls,
            operateFlowMatrixCall,
            ...wrapCalls
        ];
        subCalls.forEach(call => {
            batch.addTransaction({
                to: call.to,
                data: call.data,
                value: 0n
            });
        });
    }
    async directTransfer(to, amount, tokenAddress, txData) {
        const tokenInf = await this.sdk.data.getTokenInfo(tokenAddress);
        console.log(`Direct transfer - of: ${amount} - tokenId: ${tokenInf?.token} - to: ${to}`);
        if (!tokenInf) {
            throw new Error('Token not found');
        }
        const erc1155Types = new Set(['CrcV2_RegisterHuman', 'CrcV2_RegisterGroup']);
        const erc20Types = new Set(['CrcV2_ERC20WrapperDeployed_Demurraged', 'CrcV2_ERC20WrapperDeployed_Inflationary', 'CrcV1_Signup']);
        if (erc1155Types.has(tokenInf.type)) {
            return await this.transferErc1155(tokenAddress, to, amount, txData);
        }
        else if (erc20Types.has(tokenInf.type)) {
            return await this.transferErc20(to, amount, tokenAddress);
        }
        throw new Error(`Token type ${tokenInf.type} not supported`);
    }
    async transferErc20(to, amount, tokenAddress) {
        const iface = new ethers.Interface(['function transfer(address to, uint256 value)']);
        const data = iface.encodeFunctionData('transfer', [to, amount]);
        if (!this.sdk?.contractRunner?.sendTransaction) {
            throw new Error('ContractRunner not available');
        }
        return await this.sdk.contractRunner.sendTransaction({
            to: tokenAddress,
            data: data,
            value: 0n
        });
    }
    async transferErc1155(tokenAddress, to, amount, txData) {
        const numericTokenId = addressToUInt256(tokenAddress);
        txData = txData || new Uint8Array(0);
        console.log(`numericTokenId: ${numericTokenId}`);
        const tx = await this.sdk.v2Hub?.safeTransferFrom(this.address, to, numericTokenId, amount, txData);
        const receipt = await tx?.wait();
        if (!receipt) {
            throw new Error('Transfer failed');
        }
        return receipt;
    }
    async transfer(to, amount, tokenAddress, txData, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers) {
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        if (!tokenAddress) {
            // Detect tokens unwrap attempts and trigger unwrap execution directly
            if (this.address.toLowerCase() === to.toLowerCase() &&
                fromTokens?.length === 1 &&
                toTokens?.length === 1) {
                const [tokenFrom, tokenTo] = await Promise.all([
                    this.sdk.data.getTokenInfo(fromTokens[0]),
                    this.sdk.data.getTokenInfo(toTokens[0])
                ]);
                if (tokenFrom?.tokenOwner === tokenTo?.tokenOwner &&
                    tokenTo?.type === 'CrcV2_RegisterHuman') {
                    if (tokenFrom?.type === 'CrcV2_ERC20WrapperDeployed_Inflationary') {
                        return await this.unwrapInflationErc20(tokenFrom?.token, amount);
                    }
                    else if (tokenFrom?.type === 'CrcV2_ERC20WrapperDeployed_Demurraged') {
                        return await this.unwrapDemurrageErc20(tokenFrom?.token, amount);
                    }
                }
            }
            const batch = this.sdk.contractRunner.sendBatchTransaction();
            await this.transitiveTransfer(to, amount, batch, txData, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers);
            return (await batch.run());
        }
        else {
            return this.directTransfer(to, amount, tokenAddress, txData);
        }
    }
    async trust(avatar) {
        this.throwIfV2IsNotAvailable();
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        const avatars = Array.isArray(avatar) ? avatar : [avatar];
        const batch = this.sdk.contractRunner.sendBatchTransaction();
        for (const av of avatars) {
            const txData = this.sdk.v2Hub.interface.encodeFunctionData('trust', [av, BigInt('79228162514264337593543950335')]);
            batch.addTransaction({
                to: this.sdk.circlesConfig.v2HubAddress,
                data: txData,
                value: 0n
            });
        }
        const receipt = await batch.run();
        if (!receipt) {
            throw new Error('Trust failed');
        }
        return receipt;
    }
    async untrust(avatar) {
        this.throwIfV2IsNotAvailable();
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        const avatars = Array.isArray(avatar) ? avatar : [avatar];
        const batch = this.sdk.contractRunner.sendBatchTransaction();
        for (const av of avatars) {
            const txData = this.sdk.v2Hub.interface.encodeFunctionData('trust', [av, BigInt('0')]);
            batch.addTransaction({
                to: this.sdk.circlesConfig.v2HubAddress,
                data: txData,
                value: 0n
            });
        }
        const receipt = await batch.run();
        if (!receipt) {
            throw new Error('Untrust failed');
        }
        return receipt;
    }
    /**
     * @dev This function enables users to convert personal tokens to group tokens (gCRC)
     *      For BaseGroup, it sends tokens to the group's BaseMintHandler contract which
     *      handles the conversion process. For default group types, it calls the Hub's groupMint directly.
     *
     *      The BaseMintHandler works as follows:
     *      1. It receives ERC1155 tokens via safeBatchTransferFrom
     *      2. In its onERC1155BatchReceived/onERC1155Received functions, it:
     *         - Calls the Hub's groupMint function
     *         - Hub mints the gCRC tokens and sends them back to the BaseMintHandler
     *         - BaseMintHandler forwards tokens to the beneficiary (or wraps to ERC20 if requested)
     *
     *      The optional `data` parameter can be set to `keccak256("TYPE_DEMURRAGE")` or `keccak256("TYPE_INFLATIONARY")` constants
     *      to wrap the minted gCRC into specialized ERC20 tokens before returning to the beneficiary.
     *
     * @param group The address of the group
     * @param collateral An array of collateral token addresses to convert into group tokens
     * @param amounts An array of amounts to convert, corresponding to each collateral address
     * @param data Additional data, for BaseGroup can contain token type constants to request ERC20 wrapping
     * @returns A promise resolving to the transaction receipt after mining
     */
    async groupMint(group, collateral, amounts, data) {
        this.throwIfV2IsNotAvailable();
        group = group.toLowerCase();
        const groupType = await this.sdk.getGroupType(group);
        if (groupType == 'CrcV2_BaseGroupCreated') {
            const baseGroup = BaseGroup__factory.connect(group, this.sdk.contractRunner);
            // Get Base Group Mint handler address
            const baseGroupMintHandler = (await baseGroup.BASE_MINT_HANDLER()) ?? ZeroAddress;
            // Convert collateral tokens addresses to the uint format
            const collateralIds = collateral.map(collateralId => addressToUInt256(collateralId));
            // Send tokens to the mint handler
            const tx = await this.sdk.v2Hub.safeBatchTransferFrom(this.address, baseGroupMintHandler, collateralIds, amounts, data);
            // Wait on the receipt
            const receipt = await tx.wait();
            if (!receipt) {
                throw new Error('Group mint failed');
            }
            return receipt;
        }
        else {
            // For the default groups follow use the regular hub groupMint function
            const tx = await this.sdk.v2Hub.groupMint(group, collateral, amounts, data);
            const receipt = await tx.wait();
            if (!receipt) {
                throw new Error('Group mint failed');
            }
            return receipt;
        }
    }
    async getRedeemableAmount(group, collateral) {
        // Define the group treasury address
        const treasuryAddress = (await this.sdk.v2Hub.treasuries(group)).toLowerCase();
        return (await this.sdk.v2Hub?.balanceOf(treasuryAddress, BigInt(collateral))) || 0n;
    }
    /**
     * @notice Redeems collateral tokens from a group's treasury in exchange for group tokens
     * @dev Allows redemption of collateral from group treasuries using equivalent group tokens.
     * Implementation varies by group type:
     * - For CrcV2_BaseGroupCreated: Uses flowMatrix operations with vertex coordination
     * - For standard groups: Uses ERC1155 safeTransferFrom with encoded metadata
     *
     * @param group The address of the group from which to redeem collateral
     * @param collaterals Array of collateral token addresses to redeem
     * @param amounts Array of collateral amounts to redeem
     *
     * @return A Promise resolving to the transaction receipt upon successful redemption
     */
    async groupRedeem(group, collaterals, amounts) {
        this.throwIfV2IsNotAvailable();
        group = group.toLowerCase();
        const groupType = await this.sdk.getGroupType(group);
        if (groupType == 'CrcV2_BaseGroupCreated') {
            if (collaterals.length !== amounts.length) {
                throw new Error('Collateral and amounts arrays must be the same length');
            }
            if (!collaterals.length || !amounts.length) {
                throw new Error('Collateral and amounts arrays cannot be empty');
            }
            const totalAmount = amounts.reduce((sum, current) => sum + current, 0n);
            // Define the group treasury address
            const treasuryAddress = (await this.sdk.v2Hub.treasuries(group)).toLowerCase();
            // Address of the redeemer
            const currentAvatar = this.address.toLowerCase();
            // Check if the recipient trusts all collaterals
            for (const collateral of collaterals) {
                const isTrusted = await this.trusts(collateral);
                if (!isTrusted) {
                    throw new Error(`Collateral ${collateral} is not trusted`);
                }
            }
            if (!this.sdk?.contractRunner?.sendBatchTransaction) {
                throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
            }
            const batch = this.sdk.contractRunner.sendBatchTransaction();
            // Check if the account is approved as operator
            const approvalStatus = await this.sdk.v2Hub.isApprovedForAll(this.address, this.address);
            if (!approvalStatus) {
                const tx = this.sdk.v2Hub.interface.encodeFunctionData('setApprovalForAll', [this.address, true]);
                batch.addTransaction({
                    to: this.sdk.circlesConfig.v2HubAddress,
                    data: tx,
                    value: 0n
                });
            }
            const flowVertices = [
                // Convert to a Set to remove duplicates
                ...new Set([
                    // Construct the unsorted flow vertices array
                    ...collaterals,
                    currentAvatar,
                    group,
                    treasuryAddress
                ].map(address => address.toLowerCase()))
            ].sort((a, b) => {
                // Sort addresses in ascending order based on their numeric value
                const aValue = BigInt(a);
                const bValue = BigInt(b);
                if (aValue < bValue)
                    return -1;
                if (aValue > bValue)
                    return 1;
                return 0;
            });
            // Construct the flow array
            const flow = [{ streamSinkId: 0, amount: totalAmount.toString() }];
            const flowEdgeIds = [];
            amounts.forEach(amount => {
                flow.push({
                    streamSinkId: 1,
                    amount: amount.toString()
                });
                flowEdgeIds.push(flowEdgeIds.length + 1);
            });
            const sourceCoordinate = flowVertices.indexOf(currentAvatar);
            const groupTokenIndex = flowVertices.indexOf(group);
            const treasuryIndex = flowVertices.indexOf(treasuryAddress);
            // Construct the streams array
            const streams = [
                {
                    sourceCoordinate, // Points to sender
                    flowEdgeIds,
                    data: '0x'
                }
            ];
            const coordinates = [
                groupTokenIndex, // token
                sourceCoordinate, // from
                treasuryIndex // to
            ];
            collaterals.forEach((collateral) => {
                const collateralIndex = flowVertices.indexOf(collateral.toLowerCase());
                coordinates.push(collateralIndex, treasuryIndex, sourceCoordinate);
            });
            // The packed coordinates
            const packedCoordinates = '0x' + coordinates
                .map(index => index.toString(16).padStart(4, '0'))
                .join('');
            const tx = this.sdk.v2Hub.interface.encodeFunctionData('operateFlowMatrix', [
                flowVertices,
                flow,
                streams,
                packedCoordinates
            ]);
            batch.addTransaction({
                to: this.sdk.circlesConfig.v2HubAddress,
                data: tx,
                value: 0n
            });
            // Call the hub's operateFlowMatrix function with the constructed parameters
            const receipt = await batch.run();
            if (!receipt) {
                throw new Error('Group redeem failed');
            }
            return (receipt);
        }
        else {
            const standardTreasury = this.sdk.circlesConfig.standardTreasury;
            if (!standardTreasury) {
                throw new Error('No standard treasury address in config.');
            }
            // 1) Sum up requested redemption amounts
            let totalValue = 0n;
            for (const val of amounts) {
                totalValue += val;
            }
            if (totalValue === 0n) {
                throw new Error('Cannot redeem zero amount.');
            }
            // 2) Encode the "BaseRedemptionPolicy" data:
            //    struct BaseRedemptionPolicy {
            //       uint256[] redemptionIds;
            //       uint256[] redemptionValues;
            //    }
            // Convert each collateral address to its numeric ID:
            const redemptionIds = collaterals.map(collateral => addressToUInt256(collateral));
            const baseRedemptionPolicyEncoded = AbiCoder.defaultAbiCoder().encode(['tuple(uint256[] redemptionIds, uint256[] redemptionValues)'], [[redemptionIds, amounts]]);
            // 3) Encode the Metadata struct:
            //    struct Metadata {
            //       bytes32 metadataType;
            //       bytes   metadata;         // must be empty for groupRedeem
            //       bytes   erc1155UserData;  // your redemption policy
            //    }
            const metadataTuple = [
                this.METADATATYPE_GROUPREDEEM,
                '0x', // empty 'metadata' for groupRedeem
                baseRedemptionPolicyEncoded
            ];
            // Encode as a single tuple
            const metadataEncoded = AbiCoder.defaultAbiCoder().encode(['tuple(bytes32, bytes, bytes)'], [metadataTuple]);
            // 4) safeTransferFrom(...) to StandardTreasury
            const groupId = addressToUInt256(group);
            const tx = await this.sdk.v2Hub.safeTransferFrom(this.address, standardTreasury, groupId, totalValue, metadataEncoded);
            // 5) Wait on the receipt
            const receipt = await tx.wait();
            if (!receipt) {
                throw new Error('Group redeem failed');
            }
            return receipt;
        }
    }
    /**
     * @notice Automatically redeems collateral tokens from a Base Group's treasury
     * @dev Performs automatic redemption by determining trusted collaterals and using pathfinder for optimal flow.
     *
     * Only supports CrcV2_BaseGroupCreated group types. The function uses the v2 pathfinder to determine
     * the optimal redemption path and validates that sufficient liquidity exists before attempting redemption.
     *
     * @param group The address of the Base Group from which to redeem collateral tokens
     * @param amount The amount of group tokens to redeem for collateral (must be > 0 and <= max redeemable)
     *
     * @return A Promise resolving to the transaction receipt upon successful automatic redemption
     *
     */
    async groupRedeemAuto(group, amount) {
        this.throwIfV2IsNotAvailable();
        group = group.toLowerCase();
        const groupType = await this.sdk.getGroupType(group);
        if (groupType !== 'CrcV2_BaseGroupCreated')
            throw new Error('Only Base Groups support this method');
        // Address of the redeemer
        const currentAvatar = this.address.toLowerCase();
        // Define the group treasury address
        const treasuryAddress = (await this.sdk.v2Hub.treasuries(group)).toLowerCase();
        // Get list of all tokens in the treasury
        const treasuryTokens = (await this.sdk.data.getTokenBalances(treasuryAddress))
            .filter(balance => balance.isErc1155)
            .map(balance => balance.tokenAddress);
        const trustRelationships = await this.sdk.data.getAggregatedTrustRelations(currentAvatar, 2);
        // Get list of tokens to expect from pathfinder
        const expectedToTokens = trustRelationships.filter(trustObject => {
            if ((trustObject.relation === 'mutuallyTrusts' || trustObject.relation === 'trusts') &&
                treasuryTokens.includes(trustObject.objectAvatar))
                return true;
        }).map(trustObject => trustObject.objectAvatar);
        // Check if enough tokens as amount
        const getMaxRedeemableAmount = await this.sdk.v2Pathfinder.getMaxFlow(currentAvatar, currentAvatar, false, [group], expectedToTokens);
        if (BigInt(getMaxRedeemableAmount) < amount)
            throw new Error(`Specified amount ${amount} exceeds max tokens flow ${getMaxRedeemableAmount}`);
        const receipt = await this.transfer(currentAvatar, amount, undefined, undefined, false, [group], expectedToTokens);
        if (!receipt) {
            throw new Error('Group redeem failed');
        }
        return receipt;
    }
    async getProfile() {
        const profileCid = this.avatarInfo?.cidV0;
        if (this._cachedProfile && this._cachedProfileCid === profileCid) {
            return this._cachedProfile;
        }
        if (profileCid) {
            try {
                const profileData = await this.sdk?.profiles?.get(profileCid);
                if (profileData) {
                    this._cachedProfile = profileData;
                    this._cachedProfileCid = profileCid;
                    return this._cachedProfile;
                }
            }
            catch (e) {
                console.warn(`Couldn't load profile for CID ${profileCid}`, e);
            }
        }
        return undefined;
    }
    async registerShortNameWithNonce(nonce) {
        const tx = await this.sdk.nameRegistry?.registerShortNameWithNonce(nonce);
        if (!tx) {
            throw new Error('Failed to register short name with nonce');
        }
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Registering short name with nonce failed');
        }
        return receipt;
    }
    async updateProfile(profile) {
        const result = await this.sdk?.profiles?.create(profile);
        if (!result) {
            throw new Error('Failed to update profile. The profile service did not return a CID.');
        }
        const updateCidResult = await this.updateMetadata(result);
        if (!updateCidResult) {
            throw new Error('Failed to update profile. The CID was not updated.');
        }
        this.avatarInfo.cidV0 = result;
        return result;
    }
    async wrapDemurrageErc20(avatarAddress, amount) {
        const wrapResult = await this.sdk.v2Hub?.wrap(avatarAddress, amount, 0n /*Demurrage*/);
        const receipt = await wrapResult?.wait();
        console.log(`wrapDemurrageErc20 receipt: ${receipt}`);
        if (!receipt) {
            throw new Error('Wrap failed');
        }
        // TODO: Return the address of the wrapper
        //return await this.decodeErc20WrapperDeployed(receipt);
        return ZeroAddress;
    }
    async wrapInflationErc20(avatarAddress, amount) {
        const wrapResult = await this.sdk.v2Hub?.wrap(avatarAddress, amount, 1n /*Inflation*/);
        const receipt = await wrapResult?.wait();
        console.log(`wrapInflationErc20 receipt: ${receipt}`);
        if (!receipt) {
            throw new Error('Wrap failed');
        }
        // TODO: Return the address of the wrapper
        //return await this.decodeErc20WrapperDeployed(receipt);
        return ZeroAddress;
    }
    async unwrapDemurrageErc20(wrapperTokenAddress, amount) {
        const demurragedWrapper = await this.sdk.getDemurragedWrapper(wrapperTokenAddress);
        const tx = await demurragedWrapper.unwrap(amount);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Unwrap failed');
        }
        return receipt;
    }
    async unwrapInflationErc20(wrapperTokenAddress, amount) {
        const inflationWrapper = await this.sdk.getInflationaryWrapper(wrapperTokenAddress);
        const tx = await inflationWrapper.unwrap(amount);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Unwrap failed');
        }
        return receipt;
    }
    /**
     * Invite a user to Circles.
     * @param avatar The address of the avatar to invite. Can be either a v1 address or an address that's not signed up yet.
     */
    async inviteHuman(avatar) {
        this.throwIfV2IsNotAvailable();
        const avatarInfo = await this.sdk.data.getAvatarInfo(avatar);
        if (avatarInfo?.version == 2) {
            throw new Error('Avatar is already a v2 avatar');
        }
        const receipt = await this.trust(avatar);
        if (!receipt) {
            throw new Error('Invite failed');
        }
        return receipt;
    }
    /**
     * Gets the total supply of either this avatar's Personal- or Group-Circles, depending on the avatar's type.
     * Returns '0' for organizations or if the avatar is not signed up at Circles.
     */
    async getTotalSupply() {
        this.throwIfV2IsNotAvailable();
        return await this.sdk.v2Hub.totalSupply(this.address);
    }
    throwIfV2IsNotAvailable() {
        if (!this.sdk.circlesConfig.v2HubAddress) {
            throw new Error('V2 is not available');
        }
    }
    throwIfNameRegistryIsNotAvailable() {
        if (!this.sdk.nameRegistry) {
            throw new Error('Name registry is not available');
        }
    }
    async getStaticWrappedTokenTotalsFromSender(senderAddress) {
        const res = await this.sdk.circlesRpc.call('circles_getTokenBalances', [senderAddress]);
        return res.result.filter(o => o.tokenType == 'CrcV2_ERC20WrapperDeployed_Inflationary');
    }
}

class CMGAvatar {
    sdk;
    get address() {
        return this.avatarInfo.avatar;
    }
    avatarInfo;
    coreMemberGroup;
    _cachedProfile;
    _cachedProfileCid;
    constructor(sdk, avatarInfo) {
        this.sdk = sdk;
        this.avatarInfo = avatarInfo;
        this.coreMemberGroup = CoreMembersGroup__factory.connect(this.address, sdk.contractRunner);
        if (this.avatarInfo.version != 2) {
            throw new Error('Avatar is not a v2 avatar');
        }
    }
    groupRedeem(group, collateral, amounts) {
        throw this.NotSupportedError();
    }
    async trust(avatar, expiry) {
        expiry = BigInt(expiry || Number.MAX_SAFE_INTEGER);
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        const avatars = Array.isArray(avatar) ? avatar : [avatar];
        const batch = this.sdk.contractRunner.sendBatchTransaction();
        for (const av of avatars) {
            const txData = this.coreMemberGroup.interface.encodeFunctionData('trust', [av, expiry]);
            batch.addTransaction({
                to: this.address,
                data: txData,
                value: 0n
            });
        }
        const receipt = await batch.run();
        if (!receipt) {
            throw new Error('Trust failed');
        }
        return receipt;
    }
    async untrust(avatar) {
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        const avatars = Array.isArray(avatar) ? avatar : [avatar];
        const batch = this.sdk.contractRunner.sendBatchTransaction();
        for (const av of avatars) {
            const txData = this.coreMemberGroup.interface.encodeFunctionData('trust', [av, BigInt('0')]);
            batch.addTransaction({
                to: this.address,
                data: txData,
                value: 0n
            });
        }
        const receipt = await batch.run();
        if (!receipt) {
            throw new Error('Untrust failed');
        }
        return receipt;
    }
    owner() {
        return this.coreMemberGroup.owner();
    }
    mintHandler() {
        return this.coreMemberGroup.mintHandler();
    }
    redemptionHandler() {
        return this.coreMemberGroup.redemptionHandler();
    }
    service() {
        return this.coreMemberGroup.service();
    }
    minimalDeposit() {
        return this.coreMemberGroup.minimalDeposit();
    }
    getMembershipConditions() {
        return this.coreMemberGroup.getMembershipConditions();
    }
    trusts(otherAvatar) {
        return this.sdk.v2Hub.isTrusted(this.address, otherAvatar);
    }
    isTrustedBy(otherAvatar) {
        return this.sdk.v2Hub.isTrusted(otherAvatar, this.address);
    }
    async updateMetadata(cid) {
        const digest = cidV0ToUint8Array(cid);
        const tx = await this.coreMemberGroup.updateMetadataDigest(digest);
        const receipt = await tx?.wait();
        if (!receipt) {
            throw new Error('Updating Metadata Digest failed');
        }
        this.avatarInfo.cidV0 = cid;
        return receipt;
    }
    async getMaxTransferableAmount(to, tokenId) {
        throw this.NotSupportedError();
    }
    async getMintableAmount() {
        throw this.NotSupportedError();
    }
    async getTotalBalance() {
        return parseFloat(await this.sdk.data.getTotalBalanceV2(this.address, true));
    }
    async getGasTokenBalance() {
        // TODO: re-implement
        // return await this.sdk.contractRunner.provider?.getBalance(this.address) ?? 0n;
        return 0n;
    }
    async setOwner(owner) {
        const tx = await this.coreMemberGroup.setOwner(owner);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating owner failed');
        }
        return receipt;
    }
    async setService(service) {
        const tx = await this.coreMemberGroup.setService(service);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating service failed');
        }
        return receipt;
    }
    async setMintHandler(mintHandler) {
        const tx = await this.coreMemberGroup.setMintHandler(mintHandler);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating group mint handler failed');
        }
        return receipt;
    }
    async setRedemptionHandler(service) {
        const tx = await this.coreMemberGroup.setRedemptionHandler(service);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating group redemption handler failed');
        }
        return receipt;
    }
    async setMinimalDeposit(minimalDeposit) {
        const tx = await this.coreMemberGroup.setMinimalDeposit(minimalDeposit);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating minimal deposit failed');
        }
        return receipt;
    }
    async setFeeCollection(feeCollection) {
        const tx = await this.coreMemberGroup.setFeeCollection(feeCollection);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating fee collection failed');
        }
        return receipt;
    }
    async getTransactionHistory(pageSize) {
        const query = this.sdk.data.getTransactionHistory(this.address, pageSize);
        await query.queryNextPage();
        return query;
    }
    async getTrustRelations() {
        return this.sdk.data.getAggregatedTrustRelations(this.address, 2);
    }
    async getBalances() {
        return await this.sdk.data.getTokenBalances(this.address);
    }
    async personalMint() {
        throw this.NotSupportedError();
    }
    async stop() {
        throw this.NotSupportedError();
    }
    async transfer(to, amount, tokenAddress) {
        throw this.NotSupportedError();
    }
    async groupMint(group, collateral, amounts, data) {
        throw this.NotSupportedError();
    }
    async getProfile() {
        const profileCid = this.avatarInfo?.cidV0;
        if (this._cachedProfile && this._cachedProfileCid === profileCid) {
            return this._cachedProfile;
        }
        if (profileCid) {
            try {
                const profileData = await this.sdk?.profiles?.get(profileCid);
                if (profileData) {
                    this._cachedProfile = profileData;
                    this._cachedProfileCid = profileCid;
                    return this._cachedProfile;
                }
            }
            catch (e) {
                console.warn(`Couldn't load profile for CID ${profileCid}`, e);
            }
        }
        return undefined;
    }
    async updateProfile(profile) {
        const result = await this.sdk?.profiles?.create(profile);
        if (!result) {
            throw new Error('Failed to update profile. The profile service did not return a CID.');
        }
        const updateCidResult = await this.updateMetadata(result);
        if (!updateCidResult) {
            throw new Error('Failed to update profile. The CID was not updated.');
        }
        this.avatarInfo.cidV0 = result;
        return result;
    }
    async wrapDemurrageErc20(avatarAddress, amount) {
        throw this.NotSupportedError();
    }
    async wrapInflationErc20(avatarAddress, amount) {
        throw this.NotSupportedError();
    }
    async unwrapDemurrageErc20(wrapperTokenAddress, amount) {
        throw this.NotSupportedError();
    }
    async unwrapInflationErc20(wrapperTokenAddress, amount) {
        throw this.NotSupportedError();
    }
    async inviteHuman(avatar) {
        throw this.NotSupportedError();
    }
    /**
     * Gets the total supply of either this avatar's Personal- or Group-Circles, depending on the avatar's type.
     * Returns '0' for organizations or if the avatar is not signed up at Circles.
     */
    async getTotalSupply() {
        this.throwIfV2IsNotAvailable();
        return await this.sdk.v2Hub.totalSupply(this.address);
    }
    throwIfV2IsNotAvailable() {
        if (!this.sdk.circlesConfig.v2HubAddress) {
            throw new Error('V2 is not available');
        }
    }
    NotSupportedError() {
        return new Error('CoreMembersGroup avatar does not support this function.');
    }
}

class BaseGroupAvatar {
    sdk;
    get address() {
        return this.avatarInfo.avatar;
    }
    avatarInfo;
    baseGroup;
    _cachedProfile;
    _cachedProfileCid;
    constructor(sdk, avatarInfo) {
        this.sdk = sdk;
        this.avatarInfo = avatarInfo;
        this.baseGroup = BaseGroup__factory.connect(this.address, sdk.contractRunner);
        if (this.avatarInfo.version != 2) {
            throw new Error('Avatar is not a v2 avatar');
        }
    }
    groupRedeem(group, collateral, amounts) {
        throw this.NotSupportedError();
    }
    async trust(avatar, expiry) {
        expiry = BigInt(expiry || Number.MAX_SAFE_INTEGER);
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        const avatars = Array.isArray(avatar) ? avatar : [avatar];
        const batch = this.sdk.contractRunner.sendBatchTransaction();
        for (const av of avatars) {
            const txData = this.baseGroup.interface.encodeFunctionData('trust', [av, expiry]);
            batch.addTransaction({
                to: this.address,
                data: txData,
                value: 0n
            });
        }
        const receipt = await batch.run();
        if (!receipt) {
            throw new Error('Trust failed');
        }
        return receipt;
    }
    async trustBatchWithConditions(members, expiry = BigInt(0)) {
        const tx = await this.baseGroup.trustBatchWithConditions(members, expiry);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Trusting batch of users failed');
        }
        return receipt;
    }
    async untrust(avatar) {
        if (!this.sdk?.contractRunner?.sendBatchTransaction) {
            throw new Error('ContractRunner (or sendBatchTransaction capability) not available');
        }
        const avatars = Array.isArray(avatar) ? avatar : [avatar];
        const batch = this.sdk.contractRunner.sendBatchTransaction();
        for (const av of avatars) {
            const txData = this.baseGroup.interface.encodeFunctionData('trust', [av, BigInt('0')]);
            batch.addTransaction({
                to: this.address,
                data: txData,
                value: 0n
            });
        }
        const receipt = await batch.run();
        if (!receipt) {
            throw new Error('Untrust failed');
        }
        return receipt;
    }
    owner() {
        return this.baseGroup.owner();
    }
    mintHandler() {
        return this.baseGroup.BASE_MINT_HANDLER();
    }
    service() {
        return this.baseGroup.service();
    }
    getMembershipConditions() {
        return this.baseGroup.getMembershipConditions();
    }
    trusts(otherAvatar) {
        return this.sdk.v2Hub.isTrusted(this.address, otherAvatar);
    }
    isTrustedBy(otherAvatar) {
        return this.sdk.v2Hub.isTrusted(otherAvatar, this.address);
    }
    async updateMetadata(cid) {
        const digest = cidV0ToUint8Array(cid);
        const tx = await this.baseGroup.updateMetadataDigest(digest);
        const receipt = await tx?.wait();
        if (!receipt) {
            throw new Error('Updating Metadata Digest failed');
        }
        this.avatarInfo.cidV0 = cid;
        return receipt;
    }
    async getMaxTransferableAmount(to, tokenId) {
        throw this.NotSupportedError();
    }
    async getMintableAmount() {
        throw this.NotSupportedError();
    }
    async getTotalBalance() {
        return parseFloat(await this.sdk.data.getTotalBalanceV2(this.address, true));
    }
    async getGasTokenBalance() {
        // TODO: re-implement
        // return await this.sdk.contractRunner.provider?.getBalance(this.address) ?? 0n;
        return 0n;
    }
    async setOwner(owner) {
        const tx = await this.baseGroup.setOwner(owner);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating owner failed');
        }
        return receipt;
    }
    async setService(service) {
        const tx = await this.baseGroup.setService(service);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating service failed');
        }
        return receipt;
    }
    async setFeeCollection(feeCollection) {
        const tx = await this.baseGroup.setFeeCollection(feeCollection);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Updating fee collection failed');
        }
        return receipt;
    }
    async setMembershipCondition(condition, enabled) {
        const tx = await this.baseGroup.setMembershipCondition(condition, enabled);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Membership condition update failed');
        }
        return receipt;
    }
    async registerShortNameWithNonce(nonce) {
        const tx = await this.baseGroup.registerShortNameWithNonce(nonce);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Registering short name with nonce failed');
        }
        return receipt;
    }
    async getTransactionHistory(pageSize) {
        const query = this.sdk.data.getTransactionHistory(this.address, pageSize);
        await query.queryNextPage();
        return query;
    }
    async getTrustRelations() {
        return this.sdk.data.getAggregatedTrustRelations(this.address, 2);
    }
    async getBalances() {
        return await this.sdk.data.getTokenBalances(this.address);
    }
    async personalMint() {
        throw this.NotSupportedError();
    }
    async stop() {
        throw this.NotSupportedError();
    }
    async transfer(to, amount, tokenAddress) {
        throw this.NotSupportedError();
    }
    async groupMint(group, collateral, amounts, data) {
        throw this.NotSupportedError();
    }
    async getProfile() {
        const profileCid = this.avatarInfo?.cidV0;
        if (this._cachedProfile && this._cachedProfileCid === profileCid) {
            return this._cachedProfile;
        }
        if (profileCid) {
            try {
                const profileData = await this.sdk?.profiles?.get(profileCid);
                if (profileData) {
                    this._cachedProfile = profileData;
                    this._cachedProfileCid = profileCid;
                    return this._cachedProfile;
                }
            }
            catch (e) {
                console.warn(`Couldn't load profile for CID ${profileCid}`, e);
            }
        }
        return undefined;
    }
    async updateProfile(profile) {
        const result = await this.sdk?.profiles?.create(profile);
        if (!result) {
            throw new Error('Failed to update profile. The profile service did not return a CID.');
        }
        const updateCidResult = await this.updateMetadata(result);
        if (!updateCidResult) {
            throw new Error('Failed to update profile. The CID was not updated.');
        }
        this.avatarInfo.cidV0 = result;
        return result;
    }
    async wrapDemurrageErc20(avatarAddress, amount) {
        throw this.NotSupportedError();
    }
    async wrapInflationErc20(avatarAddress, amount) {
        throw this.NotSupportedError();
    }
    async unwrapDemurrageErc20(wrapperTokenAddress, amount) {
        throw this.NotSupportedError();
    }
    async unwrapInflationErc20(wrapperTokenAddress, amount) {
        throw this.NotSupportedError();
    }
    async inviteHuman(avatar) {
        throw this.NotSupportedError();
    }
    /**
     * Gets the total supply of either this avatar's Personal- or Group-Circles, depending on the avatar's type.
     * Returns '0' for organizations or if the avatar is not signed up at Circles.
     */
    async getTotalSupply() {
        this.throwIfV2IsNotAvailable();
        return await this.sdk.v2Hub.totalSupply(this.address);
    }
    throwIfV2IsNotAvailable() {
        if (!this.sdk.circlesConfig.v2HubAddress) {
            throw new Error('V2 is not available');
        }
    }
    NotSupportedError() {
        return new Error('BaseGroup avatar does not support this function.');
    }
}

/**
 * An Avatar represents a user registered at Circles.
 * It provides methods to interact with the Circles protocol, such as minting, transferring and trusting other avatars.
 */
class Avatar {
    address;
    /**
     * The actual avatar implementation to use behind this facade.
     * @private
     */
    _avatar;
    _avatarInfo;
    _sdk;
    /**
     * After initialization, this property contains the avatar's basic information.
     */
    get avatarInfo() {
        return this._avatarInfo;
    }
    _tokenEventSubscription = undefined;
    _groupType;
    /**
     * Creates a new Avatar instance that controls a Circles avatar at the given address.
     * @param sdk The SDK instance to use.
     * @param avatarAddress The address of the avatar to control.
     */
    constructor(sdk, avatarAddress) {
        this.address = avatarAddress;
        this._sdk = sdk;
    }
    /**
     * The events observable for this avatar.
     */
    get events() {
        if (!this._events) {
            throw new Error('Not initialized');
        }
        return this._events;
    }
    _events;
    unsubscribeFromEvents = () => {
        if (this._tokenEventSubscription) {
            this._tokenEventSubscription();
        }
    };
    subscribeToEvents = async () => {
        if (!this._avatarInfo) {
            throw new Error('Avatar is not initialized');
        }
        this._events = await this._sdk.data.subscribeToEvents(this._avatarInfo.avatar);
    };
    /**
     * Initializes the avatar.
     */
    initialize = async (subscribe = true) => {
        this.unsubscribeFromEvents();
        this._avatarInfo = await this._sdk.data.getAvatarInfo(this.address);
        if (!this._avatarInfo) {
            throw new Error('Avatar is not signed up at Circles');
        }
        const { version, hasV1 } = this._avatarInfo;
        const v1Person = () => new V1Avatar(this._sdk, this._avatarInfo);
        const v2Person = () => new V2Avatar(this._sdk, this._avatarInfo);
        const CMGroup = () => new CMGAvatar(this._sdk, this._avatarInfo);
        const BaseGroup = () => new BaseGroupAvatar(this._sdk, this._avatarInfo);
        this._groupType = await this._sdk.getGroupType(this.address);
        switch (version) {
            case 1:
                this._avatar = v1Person();
                break;
            case 2:
                if (this._groupType === 'CrcV2_CMGroupCreated') {
                    this._avatar = CMGroup();
                }
                else if (this._groupType === 'CrcV2_BaseGroupCreated') {
                    this._avatar = BaseGroup();
                }
                else if (!hasV1) {
                    this._avatar = v2Person();
                }
                else {
                    const v1Avatar = v1Person();
                    const isHuman = await v1Avatar.avatarInfo.isHuman;
                    // Handle edge case: organization migrated to v2 but still have v1 account
                    // without token which is recognized as `isStopped == false`
                    let isStopped = false;
                    if (isHuman) {
                        isStopped = await v1Avatar.v1Token?.stopped() || false;
                        this._avatar = isStopped ? v2Person() : v1Person();
                    }
                    else {
                        this._avatar = v2Person();
                    }
                    const avatarInfo = this._avatar.avatarInfo;
                    if (avatarInfo) {
                        avatarInfo.v1Stopped = isStopped;
                    }
                }
                break;
            default:
                throw new Error('Unsupported avatar');
        }
        if (subscribe) {
            await this.subscribeToEvents();
        }
    };
    onlyIfInitialized(func) {
        if (!this._avatar) {
            throw new Error('Avatar is not initialized');
        }
        return func();
    }
    onlyIfV2(func) {
        if (!this._avatar || this._avatarInfo?.version !== 2) {
            throw new Error('Avatar is not initialized or is not a v2 avatar');
        }
        return func(this._avatar);
    }
    onlyIfCoreMembersGroup(func) {
        if (!this._avatar || this._avatarInfo?.version !== 2 || this._groupType !== 'CrcV2_CMGroupCreated') {
            throw new Error('CoreMembersGroup avatar is not initialized or is not a v2 avatar');
        }
        return func(this._avatar);
    }
    onlyIfBaseGroup(func) {
        if (!this._avatar || this._avatarInfo?.version !== 2 || this._groupType !== 'CrcV2_BaseGroupCreated') {
            throw new Error('BaseGroup avatar is not initialized or is not a v2 avatar');
        }
        return func(this._avatar);
    }
    onlyIfGroup(func) {
        if (!this._avatar || this._avatarInfo?.version !== 2 || (this._groupType !== 'CrcV2_CMGroupCreated' && this._groupType !== 'CrcV2_BaseGroupCreated')) {
            throw new Error('Group avatar is not initialized or is not a v2 avatar');
        }
        return func(this._avatar);
    }
    /**
     * `human` avatars can mint 24 personal Circles per day. This method returns the amount of Circles that can be minted.
     *
     * Note: v2 avatars can mint at max. 14 days * 24 Circles = 336 Circles.
     *       v1 avatars on the other hand will stop minting after 90 days without minting.
     * @returns The amount of Circles that can be minted.
     */
    getMintableAmount = () => this.onlyIfInitialized(() => this._avatar.getMintableAmount());
    /**
     * Mints the available personal Circles for the avatar. Check `getMintableAmount()` to see how many Circles can be minted.
     * @returns The transaction receipt.
     */
    personalMint = () => this.onlyIfInitialized(() => this._avatar.personalMint());
    /**
     * Stops the avatar's token. This will prevent any future `personalMint()` calls and is not reversible.
     */
    stop = () => this.onlyIfInitialized(() => this._avatar.stop());
    /**
     * Calculates the maximum Circles amount that can be transferred to another avatar.
     *
     * NOTE: This operation can be long-running.
     * NOTE: The max. transferable amount can be lower than the avatar's balance depending on its trust relations and token holdings.
     *       Use the `getMaxTransferableAmount()` method to calculate the max. transferable amount if you need to know it beforehand.
     *
     *
     * @param to The address of the avatar to transfer to.
     * @param tokenId The token to transfer (address). Leave empty to allow transitive transfers.
     * @param useWrappedBalances If wrapped Circles should be considered in the transfers.
     * @param fromTokens If specified, makes sure that only the given tokens are used at the source.
     * @param toTokens If specified, makes sure that only the given tokens arrive at the sink.
     * @param excludeFromTokens If specified, makes sure that the given tokens are not used at the source.
     * @param excludeToTokens If specified, makes sure that the given tokens are not used at the sink.
     * @param maxTransfers The maximum number of transfers to include in the path (default: 300, ~12 MGas).
     * @returns The maximum amount that can be transferred.
     */
    getMaxTransferableAmount = (to, tokenId, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers) => this.onlyIfInitialized(() => {
        // V2 avatars support all parameters
        if (this._avatarInfo?.version === 2) {
            return this._avatar.getMaxTransferableAmount(to, tokenId, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers);
        }
        // V1 avatars only support to and tokenId
        return this._avatar.getMaxTransferableAmount(to, tokenId);
    });
    transfer(to, amount, token, txData, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers) {
        if (typeof amount === 'number') {
            if (this.avatarInfo?.version === 1) {
                const sendAttoCircles = CirclesConverter.circlesToAttoCircles(amount);
                const sendAttoCrc = CirclesConverter.attoCirclesToAttoCrc(sendAttoCircles, BigInt(Math.floor(Date.now() / 1000)));
                // V1 avatars only support basic parameters
                return this.onlyIfInitialized(() => this._avatar.transfer(to, sendAttoCrc, token, txData));
            }
            else {
                const sendAttoCircles = CirclesConverter.circlesToAttoCircles(amount);
                // V2 avatars support all parameters
                return this.onlyIfInitialized(() => this._avatar.transfer(to, sendAttoCircles, token, txData, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers));
            }
        }
        // For bigint amounts
        if (this.avatarInfo?.version === 1) {
            // V1 avatars only support basic parameters
            return this.onlyIfInitialized(() => this._avatar.transfer(to, amount, token, txData));
        }
        else {
            // V2 avatars support all parameters
            return this.onlyIfInitialized(() => this._avatar.transfer(to, amount, token, txData, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers));
        }
    }
    /**
     * Trusts another avatar. Trusting an avatar means you're willing to accept Circles that have been issued by this avatar.
     *
     * This method has two modes:
     * - Basic trust: Pass only `avatarAddress` to trust the specified avatar(s)
     * - Timed trust: (Only for base group or core members group avatars) Include `expiry` to set a time limit on trust
     *
     * @param avatarAddress The address of the avatar to trust. Can be a single address or an array of addresses.
     * @param expiry (Optional) The expiration time of the trust relationship in Unix timestamp format.
     * @returns TransactionResponse
     */
    trust(avatarAddress, expiry) {
        if (this._groupType === 'CrcV2_BaseGroupCreated' || this._groupType === 'CrcV2_CMGroupCreated') {
            return this.onlyIfGroup((avatar) => avatar.trust(avatarAddress, expiry));
        }
        return this.onlyIfInitialized(() => this._avatar.trust(avatarAddress));
    }
    /**
     * Revokes trust from another avatar. This means you will no longer accept Circles issued by this avatar. This will not affect already received Circles.
     * @param avatar The address of the avatar to untrust.
     * @returns The transaction receipt.
     */
    untrust = (avatar) => this.onlyIfInitialized(() => this._avatar.untrust(avatar));
    /**
     * Can be used to check if this avatar trusts the other avatar.
     * @param otherAvatar The address of the other avatar.
     * @return `true` if this avatar trusts the other avatar.
     */
    trusts = (otherAvatar) => this.onlyIfInitialized(() => this._avatar.trusts(otherAvatar));
    /**
     * Can be used to check if this avatar is trusted by the other avatar.
     * @param otherAvatar The address of the other avatar.
     * @return `true` if this avatar is trusted by the other avatar.
     */
    isTrustedBy = (otherAvatar) => this.onlyIfInitialized(() => this._avatar.isTrustedBy(otherAvatar));
    /**
     * Gets the trust relations of the avatar.
     * @returns An array of trust relations in this form: avatar1 - [trusts|trustedBy|mutuallyTrusts] -> avatar2.
     */
    getTrustRelations = () => this.onlyIfInitialized(() => this._avatar.getTrustRelations());
    /**
     * Gets the Circles transaction history of the avatar. The history contains incoming/outgoing transactions and minting of personal Circles and Group Circles.
     * @param pageSize The maximum number of transactions per page.
     * @returns A query object that can be used to iterate over the transaction history.
     */
    getTransactionHistory = (pageSize) => this.onlyIfInitialized(() => this._avatar.getTransactionHistory(pageSize));
    /**
     * Gets the avatar's total Circles balance.
     *
     * Note: This queries either the v1 or the v2 balance of an avatar. Check the `avatarInfo` property to see which version your avatar uses.
     *       Token holdings in v1 can be migrated to v2. Check out `Sdk.migrateAvatar` or `Sdk.migrateV1Tokens` for more information.
     */
    getTotalBalance = () => this.onlyIfInitialized(() => this._avatar.getTotalBalance());
    /**
     * Retrieves the token balances.
     *
     * This asynchronous function fetches the token balances available for the current context.
     * It requires the system to be initialized before it can be invoked.
     *
     * @returns {Promise<TokenBalanceRow[]>} A promise that resolves to an array of token balance rows.
     */
    getBalances = () => this.onlyIfInitialized(() => this._avatar.getBalances());
    /**
     * Gets the avatar's total balance of chain-native tokens.
     */
    getGasTokenBalance = () => this.onlyIfInitialized(() => this._avatar.getGasTokenBalance());
    /**
     * Use collateral, trusted by the group, to mint new Group Circles.
     * @param group The group which Circles to mint.
     * @param collateral The collateral tokens to use for minting.
     * @param amounts The amounts of the collateral tokens to use.
     * @param data Additional data for the minting operation.
     * @returns The transaction receipt.
     */
    groupMint = (group, collateral, amounts, data) => this.onlyIfV2((avatar) => avatar.groupMint(group, collateral, amounts, data));
    /**
     * Facilitates the redemption process for a specified group using the provided collateral and amounts.
     *
     * @param group - The address of the group for which the operation is being performed.
     * @param collaterals - An array of collateral addresses to redeem.
     * @param amounts - An array of amounts corresponding to each collateral.
     * @returns A promise that resolves to the transaction receipt of the redemption process.
     */
    groupRedeem = (group, collaterals, amounts) => this.onlyIfV2((avatar) => avatar.groupRedeem(group, collaterals, amounts));
    /**
     * Automatically redeems collateral tokens from a Base Group's treasury
     *
     * @param group The address of the Base Group from which to redeem collateral tokens
     * @param amount The amount of group tokens to redeem for collateral (must be > 0 and <= max redeemable)
     * @return A Promise resolving to the transaction receipt upon successful automatic redemption
     */
    groupRedeemAuto = (group, amount) => {
        return this.onlyIfV2((avatar) => {
            if (!avatar.groupRedeemAuto) {
                throw new Error('groupRedeemAuto method is not implemented');
            }
            return avatar.groupRedeemAuto(group, amount);
        });
    };
    /**
     * Wraps the specified amount of personal Circles into demurraged ERC20 tokens for use outside the Circles protocol.
     * Note: This kind of token can be incompatible with services since it's demurraged and thus the balance changes over time.
     * @param avatarAddress The address of the avatar whose Circles should be wrapped.
     * @param amount The amount of Circles to wrap.
     */
    wrapDemurrageErc20 = (avatarAddress, amount) => this.onlyIfV2((avatar) => avatar.wrapDemurrageErc20(avatarAddress, amount));
    /**
     * Wraps the specified amount of inflation Circles into ERC20 tokens for use outside the Circles protocol.
     * In contrast to demurraged tokens, these token's balance does not change over time.
     * @param avatarAddress The address of the avatar whose Circles should be wrapped.
     * @param amount The amount of Circles to wrap.
     */
    wrapInflationErc20 = (avatarAddress, amount) => this.onlyIfV2((avatar) => avatar.wrapInflationErc20(avatarAddress, amount));
    /**
     * Unwraps the specified amount of demurraged ERC20 Circles back to personal Circles.
     * @param tokenAddress The token address of the ERC20 Circles.
     * @param amount The amount of ERC20 Circles to unwrap.
     */
    unwrapDemurrageErc20 = (tokenAddress, amount) => this.onlyIfV2((avatar) => avatar.unwrapDemurrageErc20(tokenAddress, amount));
    /**
     * Unwraps the specified amount of inflation ERC20 Circles back to personal Circles.
     * @param avatarAddress The address of the avatar whose Circles should be unwrapped.
     * @param amount The amount of ERC20 Circles to unwrap.
     */
    unwrapInflationErc20 = (avatarAddress, amount) => this.onlyIfV2((avatar) => avatar.unwrapInflationErc20(avatarAddress, amount));
    /**
     * Invite a human avatar to join Circles.
     * @param avatar The address of any human controlled wallet.
     */
    inviteHuman = (avatar) => this.onlyIfV2((_avatar) => _avatar.inviteHuman(avatar));
    /**
     * Updates the avatar's metadata (profile).
     * @param cid The IPFS content identifier of the metadata (Qm....).
     * @returns The transaction receipt confirming the update.
     */
    updateMetadata = (cid) => this.onlyIfInitialized(() => this._avatar.updateMetadata(cid));
    /**
     * Gets the profile that's associated with the avatar or returns `undefined` if no profile is associated.
     * @returns The profile or `undefined`.
     */
    getProfile = () => this.onlyIfV2((_avatar) => _avatar.getProfile());
    /**
     * Updates the avatar's profile.
     * @param profile The new profile.
     * @returns The IPFS CID of the updated profile.
     */
    updateProfile = (profile) => this.onlyIfV2((_avatar) => _avatar.updateProfile(profile));
    /**
     * Gets the total supply of either this avatar's Personal or Group Circles.
     */
    getTotalSupply = () => this.onlyIfInitialized(() => this._avatar.getTotalSupply());
    /// Methods for CMGAvatar
    /**
     * Retrieves the owner of the group.
     * @returns The address of the owner.
     */
    owner = () => this.onlyIfGroup((avatar) => avatar.owner());
    /**
     * Retrieves the mint handler address, responsible for minting new group tokens.
     * @returns The address of the mint handler.
     */
    mintHandler = () => this.onlyIfGroup((avatar) => avatar.mintHandler());
    /**
     * Retrieves the redemption handler address, responsible for handling group token redemptions.
     * @returns The address of the redemption handler.
     */
    redemptionHandler = () => this.onlyIfCoreMembersGroup((avatar) => avatar.redemptionHandler());
    /**
     * Retrieves the service address associated with the group contract.
     * @returns The address of the service.
     */
    service = () => this.onlyIfGroup((avatar) => avatar.service());
    /**
     * Gets the minimum deposit required for group mint operation.
     * @returns The minimum deposit amount as a `bigint`.
     */
    minimalDeposit = () => this.onlyIfCoreMembersGroup((avatar) => avatar.minimalDeposit());
    /**
     * Gets the membership conditions required to enter the group.
     * @returns Set of addresses representing membership conditions.
     */
    getMembershipConditions = () => this.onlyIfGroup((avatar) => avatar.getMembershipConditions());
    /**
     * Updates the owner of the group to a new address.
     * This functionality is only available for Base or Core Members Group avatars.
     *
     * @param owner The address of the new owner
     * @returns A promise resolving to the transaction receipt
     * @throws Error if the avatar is not initialized or is not a Base or Core Members Group avatar
     */
    setOwner = (owner) => this.onlyIfGroup((avatar) => avatar.setOwner(owner));
    /**
     * Sets the service address for the group contract.
     * @param service The new service address to be set.
     * @returns A promise resolving to the transaction receipt.
     */
    setService = (service) => this.onlyIfGroup((avatar) => avatar.setService(service));
    /**
     * Assigns a new mint handler responsible for mint operations.
     * @param mintHandler The address of the new mint handler.
     * @returns A promise resolving to the transaction receipt.
     */
    setMintHandler = (mintHandler) => this.onlyIfCoreMembersGroup((avatar) => avatar.setMintHandler(mintHandler));
    /**
     * Assigns a new redemption handler responsible for handling redemptions.
     * @param redemptionHandler The address of the new redemption handler.
     * @returns A promise resolving to the transaction receipt.
     */
    setRedemptionHandler = (redemptionHandler) => this.onlyIfCoreMembersGroup((avatar) => avatar.setRedemptionHandler(redemptionHandler));
    /**
     * Sets the minimal deposit required for the group mint.
     * @param minimalDeposit The amount representing the minimal deposit requirement.
     * @returns A promise resolving to the transaction receipt.
     */
    setMinimalDeposit = (minimalDeposit) => this.onlyIfCoreMembersGroup((avatar) => avatar.setMinimalDeposit(minimalDeposit));
    /**
     * Assigns a new address for fee collection.
     * @param feeCollection The address of the new fee collection entity.
     * @returns A promise resolving to the transaction receipt.
     */
    setFeeCollection = (feeCollection) => this.onlyIfGroup((avatar) => avatar.setFeeCollection(feeCollection));
    /**
     * Sets or updates a membership condition for the group.
     * This functionality is only available for Base Group avatars.
     *
     * @param condition The address representing a membership condition
     * @param enabled Boolean flag indicating whether the condition should be enabled or disabled
     * @returns A promise resolving to the transaction receipt
     * @throws Error if the avatar is not initialized or is not a Base Group avatar
     */
    setMembershipCondition = (condition, enabled) => this.onlyIfBaseGroup((avatar) => avatar.setMembershipCondition(condition, enabled));
    /**
     * Registers a short name for the group using a nonce value.
     * This functionality is only available for Base Group avatars.
     *
     * @param nonce A user-provided nonce to handle name collisions or concurrency issues
     * @returns A promise resolving to the transaction receipt
     * @throws Error if the avatar is not initialized or is not a Base Group avatar or if registration fails
     */
    registerShortNameWithNonce = (nonce) => {
        if (this._avatar instanceof BaseGroupAvatar) {
            return this._avatar.registerShortNameWithNonce(nonce);
        }
        if (this._avatar instanceof V2Avatar) {
            return this._avatar.registerShortNameWithNonce(nonce);
        }
        throw new Error('Avatar is not initialized or is not a Base Group or V2 avatar');
    };
    /**
     * Establishes trust relationships with multiple addresses in a single transaction with expiry conditions.
     * This functionality is only available for Base Group avatars.
     *
     * @param members An array of addresses to trust
     * @param expiry Optional timestamp when the trust relationships will expire (defaults to 0)
     * @returns A promise resolving to the transaction receipt
     * @throws Error if the avatar is not initialized or is not a Base Group avatar
     */
    trustBatchWithConditions = (members, expiry) => this.onlyIfBaseGroup((avatar) => avatar.trustBatchWithConditions(members, expiry));
}

class V1Pathfinder {
    rpc;
    constructor(circlesRpcUrl) {
        this.rpc = new CirclesRpc(circlesRpcUrl); // Using CirclesRpc class
    }
    async getMaxFlow(from, to) {
        const requestBody = {
            Source: from,
            Sink: to,
            TargetFlow: "99999999999999999999999999999999999999", // A large target flow
        };
        const response = await this.rpc.call('circlesV1_findPath', [requestBody]);
        return BigInt(response.result.maxFlow);
    }
    async getPath(from, to, value) {
        const requestBody = {
            Source: from,
            Sink: to,
            TargetFlow: value.toString(),
        };
        const response = await this.rpc.call('circlesV1_findPath', [requestBody]);
        return response.result;
    }
}

const circlesConfig = {
    100: {
        circlesRpcUrl: "https://rpc.aboutcircles.com/",
        pathfinderUrl: "https://pathfinder.aboutcircles.com",
        profileServiceUrl: "https://rpc.aboutcircles.com/profiles/",
        v1HubAddress: "0x29b9a7fbb8995b2423a71cc17cf9810798f6c543",
        v2HubAddress: "0xc12C1E50ABB450d6205Ea2C3Fa861b3B834d13e8",
        nameRegistryAddress: "0xA27566fD89162cC3D40Cb59c87AAaA49B85F3474",
        v1NameRegistryAddress: "0x1eaD7F904F6fFC619c58B85e04F890b394E08172",
        migrationAddress: "0xD44B8dcFBaDfC78EA64c55B705BFc68199B56376",
        baseGroupMintPolicy: "0xcCa27c26CF7BAC2a9928f42201d48220F0e3a549",
        standardTreasury: "0x08F90aB73A515308f03A718257ff9887ED330C6e",
        coreMembersGroupDeployer: "0xFEca40Eb02FB1f4F5F795fC7a03c1A27819B1Ded",
        baseGroupFactory: "0xD0B5Bd9962197BEaC4cbA24244ec3587f19Bd06d"
    }
};

class V2Pathfinder {
    rpcUrl;
    constructor(circlesRpcUrl) {
        this.rpcUrl = circlesRpcUrl;
    }
    async getMaxFlow(from, to, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers) {
        to = to.toLowerCase();
        const result = await findMaxFlow(this.rpcUrl, {
            from,
            to,
            useWrappedBalances,
            fromTokens,
            toTokens,
            excludeFromTokens,
            excludeToTokens,
            maxTransfers
        });
        return CirclesConverter.truncateToSixDecimals(result);
    }
    async getPath(from, to, value, useWrappedBalances, fromTokens, toTokens, excludeFromTokens, excludeToTokens, maxTransfers) {
        return await findPath(this.rpcUrl, {
            from,
            to,
            targetFlow: value,
            useWrappedBalances,
            fromTokens,
            toTokens,
            excludeFromTokens,
            excludeToTokens,
            maxTransfers
        });
    }
}

/**
 * The SDK provides a high-level interface to interact with the Circles protocol.
 */
class Sdk {
    /**
     * The signer used to sign transactions.
     */
    contractRunner;
    /**
     * The chain specific Circles configuration.
     */
    circlesConfig;
    /**
     * The Circles RPC client.
     */
    circlesRpc;
    /**
     * The Circles data client.
     */
    data;
    /**
     * The typechain generated V1 hub contract wrapper.
     */
    v1Hub;
    /**
     * The typechain generated V2 hub contract wrapper.
     */
    v2Hub;
    /**
     * The typechain generated NameRegistry contract wrapper.
     */
    nameRegistry;
    /**
     * The typechain generated V1NameRegistry contract wrapper.
     */
    v1NameRegistry;
    /**
     * The pathfinder client (v1).
     */
    v1Pathfinder;
    /**
     * The pathfinder client (v2).
     */
    v2Pathfinder;
    /**
     * The Core Members Group deployer (v2).
     */
    coreMembersGroupDeployer;
    /**
     * The Base Group deployer (v2).
     */
    baseGroupFactory;
    /**
     * The profiles service client.
     */
    profiles;
    /**
     * Contains the bootstrap periods for each known hub contract.
     */
    bootstrapPeriods = {
        '0xc12c1e50abb450d6205ea2c3fa861b3b834d13e8': /*deployedAt:*/ 1728824950 + /*bootsrapTime:*/ 2883058,
        '0x3d61f0a272ec69d65f5cff097212079aafde8267': /*deployedAt:*/ 1730401610 + /*bootsrapTime:*/ 1313598
    };
    /**
     * Creates a new SDK instance.
     * @param contractRunner A contract runner instance and its address.
     * @param config The optional chain specific Circles configuration.
     */
    constructor(contractRunner, config) {
        this.circlesConfig = config ?? circlesConfig[100];
        this.contractRunner = contractRunner;
        if (!this.contractRunner.address) {
            throw new Error('Contract runner is not initialized');
        }
        this.circlesRpc = new CirclesRpc(this.circlesConfig.circlesRpcUrl);
        this.v2Pathfinder = new V2Pathfinder(this.circlesConfig.circlesRpcUrl);
        this.data = new CirclesData(this.circlesRpc);
        this.v1Hub = Hub__factory.connect(this.circlesConfig.v1HubAddress ?? '0x29b9a7fBb8995b2423a71cC17cf9810798F6C543', this.contractRunner);
        if (this.circlesConfig.v2HubAddress) {
            this.v2Hub = Hub__factory$1.connect(this.circlesConfig.v2HubAddress, this.contractRunner);
        }
        if (this.circlesConfig.pathfinderUrl) {
            this.v1Pathfinder = new V1Pathfinder(this.circlesConfig.circlesRpcUrl);
        }
        if (this.circlesConfig.nameRegistryAddress) {
            this.nameRegistry = NameRegistry__factory.connect(this.circlesConfig.nameRegistryAddress, this.contractRunner);
        }
        if (this.circlesConfig.v1NameRegistryAddress) {
            this.v1NameRegistry = NameRegistryV1__factory.connect(this.circlesConfig.v1NameRegistryAddress, this.contractRunner);
        }
        if (this.circlesConfig.profileServiceUrl) {
            this.profiles = new Profiles(this.circlesConfig.profileServiceUrl);
        }
        if (this.circlesConfig.coreMembersGroupDeployer) {
            this.coreMembersGroupDeployer = CMGroupDeployer__factory.connect(this.circlesConfig.coreMembersGroupDeployer, this.contractRunner);
        }
        if (this.circlesConfig.baseGroupFactory) {
            this.baseGroupFactory = BaseGroupFactory__factory.connect(this.circlesConfig.baseGroupFactory, this.contractRunner);
        }
    }
    /**
     * Gets an avatar by its address.
     * @param avatarAddress The avatar's address.
     * @param subscribe Whether to subscribe to avatar events.
     * @returns The avatar instance.
     * @throws If the given avatar address is not signed up at Circles.
     */
    getAvatar = async (avatarAddress, subscribe = true) => {
        const avatar = new Avatar(this, avatarAddress);
        await avatar.initialize(subscribe);
        return avatar;
    };
    /**
     * Creates or updates a profile and registers its CID in the NameRegistry.
     *
     * @param {Profile | string} profile - The profile information or profile ID to be created or updated.
     * @returns {Promise<ContractTransactionReceipt>} - A promise that resolves to the transaction receipt.
     * @throws {Error} - Throws an error if the Profiles service or NameRegistry is not configured,
     *                   or if the transaction fails.
     */
    createOrUpdateProfile = async (profile) => {
        if (!this.profiles) {
            throw new Error('Profiles service is not configured');
        }
        if (!this.nameRegistry) {
            throw new Error('NameRegistry is not configured');
        }
        const profileCid = await this.createProfileIfNecessary(profile);
        // Register the profile CID in the NameRegistry
        const tx = await this.nameRegistry.updateMetadataDigest(profileCid);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Transaction failed');
        }
        return receipt;
    };
    /**
     * Registers the connected wallet as a human avatar.
     * @returns The avatar instance.
     */
    registerHuman = async () => {
        const receipt = await this.v1Hub.signup();
        await receipt.wait();
        await this.waitForAvatarInfo(this.contractRunner.address);
        return this.getAvatar(this.contractRunner.address);
    };
    async acceptInvitation(inviter, profile) {
        inviter = inviter.toLowerCase();
        return this._registerHuman(inviter, profile);
    }
    async _registerHuman(inviter, profile) {
        if (!this.v2Hub) {
            throw new Error('V2 hub not available');
        }
        let metadataDigest = await this.createProfileIfNecessary(profile);
        const tx = await this.v2Hub.registerHuman(inviter, metadataDigest);
        const receipt = await tx.wait();
        if (!receipt) {
            throw new Error('Transaction failed');
        }
        await this.waitForAvatarInfo(this.contractRunner.address);
        return this.getAvatar(this.contractRunner.address);
    }
    /**
     * Checks if the profile argument is a string or a Profile object and creates the profile if necessary.
     * If the profile is a string, it must be a CIDv0 string (Qm...).
     * @param profile The profile data or CIDv0 of the avatar.
     * @private
     */
    async createProfileIfNecessary(profile) {
        if (typeof profile === 'string') {
            if (!profile.startsWith('Qm')) {
                throw new Error('Invalid profile CID. Must be a CIDv0 string (Qm...).');
            }
            return cidV0ToUint8Array(profile);
        }
        else if (this.profiles) {
            const profileCid = await this.profiles?.create(profile);
            return cidV0ToUint8Array(profileCid);
        }
        else {
            throw new Error('Profiles service is not configured');
        }
    }
    /**
     * Registers the connected wallet as an organization avatar.
     * @returns The avatar instance.
     */
    registerOrganization = async () => {
        const receipt = await this.v1Hub.organizationSignup();
        await receipt.wait();
        await this.waitForAvatarInfo(this.contractRunner.address);
        return this.getAvatar(this.contractRunner.address);
    };
    /**
     * Registers the connected wallet as an organization avatar in Circles v2.
     * @param profile The profile data of the organization.
     */
    registerOrganizationV2 = async (profile) => {
        if (!this.v2Hub) {
            throw new Error('V2 hub not available');
        }
        const metadataDigest = await this.createProfileIfNecessary(profile);
        const receipt = await this.v2Hub.registerOrganization(profile.name, metadataDigest);
        await receipt.wait();
        await this.waitForAvatarInfo(this.contractRunner.address);
        return this.getAvatar(this.contractRunner.address);
    };
    /**
     * Registers the connected wallet as a group avatar in Circles v2.
     * @param mint The address of the minting policy contract to use.
     * @param profile The profile data of the group.
     */
    registerGroupV2 = async (mint, profile) => {
        if (!this.v2Hub) {
            throw new Error('V2 hub not available');
        }
        const metadataDigest = await this.createProfileIfNecessary(profile);
        const receipt = await this.v2Hub.registerGroup(mint, profile.name, profile.symbol, metadataDigest);
        await receipt.wait();
        await this.waitForAvatarInfo(this.contractRunner.address);
        return this.getAvatar(this.contractRunner.address);
    };
    waitForAvatarInfo = async (address) => {
        let avatarRow;
        let retries = 0;
        do {
            avatarRow = await this.data.getAvatarInfo(address);
            await new Promise((resolve) => setTimeout(resolve, 500));
            retries++;
            if (retries > 120) {
                throw new Error(`Timeout getting avatar info for ${address}`);
            }
        } while (!avatarRow);
        return avatarRow;
    };
    /**
     * Migrates a v1 avatar, and optionally it's trust relations to v2.
     * @param inviter The address of the avatar that invited the user to v2. Can be 'ZeroAddress' during migration period or if the account that's migrating stopped minting in v1 during the migration period.
     * @param avatar The avatar's address.
     * @param profile The profile data of the avatar.
     * @param trustRelations An optional list of trust relations to migrate.
     */
    migrateAvatar = async (inviter, avatar, profile, trustRelations, isRawCalldata) => {
        inviter = inviter.toLowerCase();
        avatar = avatar.toLowerCase();
        if (!this.v2Hub) {
            throw new Error('V2 hub not available');
        }
        const avatarInfo = await this.data.getAvatarInfo(avatar);
        if (!avatarInfo) {
            throw new Error('Avatar not found');
        }
        // Check if the contract runner supports batch transactions
        const contractRunner = this.contractRunner;
        if (!contractRunner.sendBatchTransaction) {
            throw new Error('Batch transaction not supported by contract runner');
        }
        // Create a new batch
        const batch = contractRunner.sendBatchTransaction();
        if (avatarInfo.hasV1) {
            // 1. Stop V1 token if necessary
            if (avatarInfo.v1Token) {
                const v1Avatar = new V1Avatar(this, avatarInfo);
                const isStopped = await v1Avatar.v1Token?.stopped();
                if (!isStopped) {
                    if (!v1Avatar.v1Token) {
                        throw new Error(`Could not load V1 token for avatar ${avatar}. Token address: ${avatarInfo.v1Token}`);
                    }
                    // Add 'personalMint' to the batch
                    const personalMintData = v1Avatar.v1Token.interface.encodeFunctionData('update');
                    const personalMintTx = {
                        to: avatarInfo.v1Token,
                        data: personalMintData,
                        value: 0n
                    };
                    batch.addTransaction(personalMintTx);
                    // Add 'v1Token.stop' to the batch
                    const stopData = v1Avatar.v1Token.interface.encodeFunctionData('stop');
                    const stopTx = {
                        to: avatarInfo.v1Token,
                        data: stopData,
                        value: 0n
                    };
                    batch.addTransaction(stopTx);
                }
            }
            // 2. Signup V2 avatar if necessary
            if (avatarInfo.version === 1) {
                // Add 'registerHumanV2' to the batch
                const metadataDigest = await this.createProfileIfNecessary(profile);
                if (avatarInfo.type === 'CrcV1_Signup') {
                    // Check if the account that's migrating stopped minting in v1 during the migration period.
                    // Throw an error otherwise.
                    if (inviter === ZeroAddress && !(await this.canSelfMigrate(avatarInfo))) {
                        throw new Error(`Self registration not allowed for avatar ${avatar} because it did not stop minting in v1 during the migration period`);
                    }
                    // Add 'registerHuman' to the batch
                    const registerHumanData = this.v2Hub.interface.encodeFunctionData('registerHuman', [inviter, metadataDigest]);
                    const registerHumanTx = {
                        to: this.circlesConfig.v2HubAddress,
                        data: registerHumanData,
                        value: 0n
                    };
                    batch.addTransaction(registerHumanTx);
                }
                else if (avatarInfo.type === 'CrcV1_OrganizationSignup') {
                    // Add 'registerOrganization' to the batch
                    const registerOrganizationData = this.v2Hub.interface.encodeFunctionData('registerOrganization', [profile.name, metadataDigest]);
                    const registerOrganizationTx = {
                        to: this.circlesConfig.v2HubAddress,
                        data: registerOrganizationData,
                        value: 0n
                    };
                    batch.addTransaction(registerOrganizationTx);
                }
                else {
                    throw new Error(`Avatar type ${avatarInfo.type} not supported`);
                }
            }
            if (avatarInfo.isHuman) {
                // 3. Ensure the v1 token minting status is known to the v2 hub
                // Add 'calculateIssuanceTx' to the batch
                const calculateIssuanceData = this.v2Hub.interface.encodeFunctionData('calculateIssuanceWithCheck', [avatar]);
                const calculateIssuanceTx = {
                    to: this.circlesConfig.v2HubAddress,
                    data: calculateIssuanceData,
                    value: 0n
                };
                batch.addTransaction(calculateIssuanceTx);
            }
            // 4. Migrate v1 balances of the avatar. But only if the tokenOwner of the migrated token is already on v2 or if it's the own token.
            const balances = await this.data.getTokenBalances(avatar);
            const v1Balances = balances.filter(o => o.version === 1);
            const tokenOwners = v1Balances.map(o => o.tokenOwner);
            const tokenOwnerInfo = await this.data.getAvatarInfoBatch(tokenOwners);
            const tokenOwnerInfoMap = tokenOwnerInfo.reduce((acc, o) => {
                acc[o.avatar] = o;
                return acc;
            }, {});
            const tokensToMigrate = v1Balances.filter(o => o.tokenOwner == avatar || tokenOwnerInfoMap[o.tokenOwner]?.version === 2);
            if (tokensToMigrate.length > 0) {
                // Add 'migrate' to the batch
                await this.migrateV1TokensBatch(avatar, tokensToMigrate.map(o => o.tokenAddress), batch);
            }
            // 5. Migrate trust relations
            if (trustRelations) {
                // Add 'trust' to the batch
                for (const trustRelation of trustRelations) {
                    const trustData = this.v2Hub.interface.encodeFunctionData('trust', [trustRelation, BigInt('79228162514264337593543950335')]);
                    const trustTx = {
                        to: this.circlesConfig.v2HubAddress,
                        data: trustData,
                        value: 0n
                    };
                    batch.addTransaction(trustTx);
                }
            }
            // Run the batch
            if (isRawCalldata) {
                if (!batch.getTxCalldata) {
                    throw new Error('Contract runner does not support raw call data');
                }
                console.log('Return raw tx calldata');
                return await batch.getTxCalldata();
            }
            else {
                const batchResponse = await batch.run();
                console.log('Batch transaction response:', batchResponse);
            }
        }
        else {
            throw new Error('Avatar is not a V1 avatar');
        }
    };
    /**
     * Checks if an avatar can self-migrate to v2.
     * This is possible either during the migration period or if the avatar stopped minting in v1 before or during that period.
     * @param avatarInfo The avatar's info.
     * @private
     */
    async canSelfMigrate(avatarInfo) {
        if (!this.circlesConfig.v2HubAddress) {
            throw new Error('V2 hub address not set');
        }
        avatarInfo.avatar = avatarInfo.avatar.toLowerCase();
        const v1Avatar = new V1Avatar(this, avatarInfo);
        const v1Token = v1Avatar.v1Token;
        if (!v1Token) {
            return false;
        }
        const ringsDeployment = '0x3d61f0a272ec69d65f5cff097212079aafde8267';
        if (this.circlesConfig.v2HubAddress.toLowerCase() === ringsDeployment) {
            return true;
        }
        const migrationPeriodEnd = this.bootstrapPeriods[this.circlesConfig.v2HubAddress.toLowerCase()];
        if (Date.now() / 1000 < migrationPeriodEnd) {
            // Allow to self migrate anyone with a v1 token during the migration period
            return true;
        }
        // Once the migration period is over, verify that minting stopped before or during the migration period
        const [isStopped, lastMint] = await Promise.all([
            v1Token.stopped(),
            v1Token.lastTouched()
        ]);
        return isStopped && lastMint <= migrationPeriodEnd;
    }
    /**
     * Migrates all V1 token holdings of an avatar to V2 using batch transactions.
     * @param avatar The avatar whose tokens to migrate.
     * @param tokens An optional list of token addresses to migrate. If not provided, all tokens will be migrated.
     * @param batch An optional batch transaction to add transactions to.
     */
    migrateV1TokensBatch = async (avatar, tokens, batch) => {
        avatar = avatar.toLowerCase();
        tokens = tokens?.map(o => o.toLowerCase());
        if (!this.circlesConfig.migrationAddress) {
            throw new Error('Migration address not set');
        }
        // If no batch is provided, create one
        let ownBatch = false;
        if (!batch) {
            const contractRunner = this.contractRunner;
            if (!contractRunner.sendBatchTransaction) {
                throw new Error('Batch transaction not supported by contract runner');
            }
            batch = contractRunner.sendBatchTransaction();
            ownBatch = true;
        }
        const tokenSet = new Set(tokens?.map(o => o) ?? []);
        const balances = await this.data.getTokenBalances(avatar);
        const v1Balances = balances.filter(o => o.version === 1 &&
            (tokenSet.size > 0 ? tokenSet.has(o.tokenAddress) : true));
        const tokensToMigrate = v1Balances.filter(o => BigInt(o.attoCrc) > 0n);
        console.log(`Migrating the following v1 tokens:`, tokensToMigrate);
        // Fetch allowances in parallel
        const allowances = await Promise.all(tokensToMigrate.map(async (tokenToMigrate) => {
            const token = Token__factory.connect(tokenToMigrate.tokenAddress, this.contractRunner);
            return await token.allowance(avatar, this.circlesConfig.migrationAddress);
        }));
        // Process tokensToMigrate and allowances
        for (let i = 0; i < tokensToMigrate.length; i++) {
            const tokenToMigrate = tokensToMigrate[i];
            const allowance = allowances[i];
            const balance = BigInt(tokenToMigrate.attoCrc);
            const token = Token__factory.connect(tokenToMigrate.tokenAddress, this.contractRunner);
            if (allowance < balance) {
                const increase = balance - allowance;
                const increaseAllowanceData = token.interface.encodeFunctionData('increaseAllowance', [this.circlesConfig.migrationAddress, increase]);
                const tx = {
                    to: tokenToMigrate.tokenAddress,
                    data: increaseAllowanceData,
                    value: 0n
                };
                batch.addTransaction(tx);
            }
        }
        // Add the migrate call
        const migrationContract = Migration__factory.connect(this.circlesConfig.migrationAddress, this.contractRunner);
        const tokensToMigrateAddresses = tokensToMigrate.map(o => o.tokenOwner);
        const amountsToMigrate = tokensToMigrate.map(o => BigInt(o.attoCrc));
        const migrateData = migrationContract.interface.encodeFunctionData('migrate', [tokensToMigrateAddresses, amountsToMigrate]);
        const migrateTx = {
            to: this.circlesConfig.migrationAddress,
            data: migrateData,
            value: 0n
        };
        batch.addTransaction(migrateTx);
        // If we created our own batch, run it
        if (ownBatch) {
            const batchResponse = await batch.run();
            console.log('Batch transaction response:', batchResponse);
        }
    };
    /**
     * Migrates all V1 token holdings of an avatar to V2.
     * @param avatar The avatar whose tokens to migrate.
     * @param tokens An optional list of token addresses to migrate. If not provided, all tokens will be migrated.
     */
    migrateV1Tokens = async (avatar, tokens) => {
        avatar = avatar.toLowerCase();
        tokens = tokens?.map(o => o.toLowerCase());
        if (!this.circlesConfig.migrationAddress) {
            throw new Error('Migration address not set');
        }
        const balances = await this.data.getTokenBalances(avatar);
        const v1Balances = balances.filter(o => o.version === 1 && (tokens ? tokens.includes(o.tokenAddress) : true));
        const tokensToMigrate = v1Balances.filter(o => BigInt(o.attoCrc) > 0);
        console.log(`Migrating the following v1 token:`, tokensToMigrate);
        // TODO: Send in one transaction if sent to Safe
        await Promise.all(tokensToMigrate.map(async (tokenToMigrate) => {
            const balance = BigInt(tokenToMigrate.attoCrc);
            console.log(`tokenToMigrate`, tokenToMigrate);
            const token = Token__factory.connect(tokenToMigrate.tokenAddress, this.contractRunner);
            const allowance = await token.allowance(avatar, this.circlesConfig.migrationAddress);
            if (allowance < balance) {
                const increase = balance - allowance;
                const tx = await token.increaseAllowance(this.circlesConfig.migrationAddress, increase);
                await tx.wait();
            }
        }));
        const migrationContract = Migration__factory.connect(this.circlesConfig.migrationAddress, this.contractRunner);
        const migrateTx = await migrationContract.migrate(tokensToMigrate.map(o => o.tokenOwner), tokensToMigrate.map(o => BigInt(o.attoCrc)));
        await migrateTx.wait();
    };
    getInflationaryWrapper = async (wrapperAddress) => {
        return InflationaryCircles__factory.connect(wrapperAddress, this.contractRunner);
    };
    getDemurragedWrapper = async (wrapperAddress) => {
        return DemurrageCircles__factory.connect(wrapperAddress, this.contractRunner);
    };
    isCoreMembersGroup = async (avatar) => {
        const results = this.data.findGroups(1, {
            groupAddressIn: [avatar],
            groupTypeIn: ['CrcV2_CMGroupCreated']
        });
        return !!(await results.getSingleRow());
    };
    getGroupType = async (avatar) => {
        const results = this.data.findGroups(1, {
            groupAddressIn: [avatar]
        });
        return (await results.getSingleRow())?.type;
    };
    /**
     * Checks if the `to` address is a group minter and excludes the group tokens from the transfer
     * if that's the case.
     * @param to The receiver of the transfer
     * @param excludeFromTokens The existing list of tokens to exclude from the transfer
     * @returns The complete list of tokens to exclude from the transfer
     */
    async getDefaultTokenExcludeList(to, excludeFromTokens) {
        const groupInfoByMintHandler = this.data.findGroups(1, {
            mintHandlerEquals: to
        });
        const groupInfo = await groupInfoByMintHandler.getSingleRow();
        const completeExcludeFromTokenList = new Set();
        if (groupInfo) {
            completeExcludeFromTokenList.add(groupInfo.group);
            if (groupInfo.erc20WrapperDemurraged) {
                completeExcludeFromTokenList.add(groupInfo.erc20WrapperDemurraged);
            }
            if (groupInfo.erc20WrapperStatic) {
                completeExcludeFromTokenList.add(groupInfo.erc20WrapperStatic);
            }
        }
        excludeFromTokens?.forEach(completeExcludeFromTokenList.add);
        if (completeExcludeFromTokenList.size == 0)
            return undefined;
        return Array.from(completeExcludeFromTokenList);
    }
}

export { Avatar, Sdk, V1Avatar, V1Pathfinder, circlesConfig };
